diff -ru a/Makefile.in b/Makefile.in
--- a/Makefile.in	Wed Apr  9 07:13:34 2025
+++ b/Makefile.in	Fri Apr 25 15:47:04 2025
@@ -28884,8 +28884,11 @@
 	  && t=$*.td							\
 	  && rm -rf $$t							\
 	  && $(MKDIR_P) $$t						\
-	  && (cd $$t && $(LN_S) '$(abs_top_builddir)/src/'$$prog$(EXEEXT) \
-				$$argv$(EXEEXT))			\
+	  && (cd $$t && (echo "#!/bin/bash" &&				\
+	  	echo "export LD_LIBRARY_PATH=`echo $(abs_top_builddir) | \
+		sed 's/projects\/illumos-extra\/coreutils\/coreutils.*-64$$/proto\/lib\/64/g'`" \
+		&& echo '$(abs_top_builddir)/src/'$$prog$(EXEEXT) '$$@' ) \
+			> $$argv$(EXEEXT) && chmod 0755 $$argv$(EXTEXT) ) \
 	&& : $${SOURCE_DATE_EPOCH=`date -r $(srcdir)/.tarball-version +%s \
 				   2>/dev/null || :`} \
 	&& : $${TZ=UTC0} && export TZ					\
