diff -ru a/g10/options.h b/g10/options.h
--- a/g10/options.h	2016-08-17 05:00:11.000000000 -0400
+++ b/g10/options.h	2021-11-17 14:56:25.000000000 -0500
@@ -27,7 +27,7 @@
 
 #ifndef EXTERN_UNLESS_MAIN_MODULE
 /* Norcraft can't cope with common symbols */
-#if defined (__riscos__) && !defined (INCLUDED_BY_MAIN_MODULE)
+#if !defined (INCLUDED_BY_MAIN_MODULE)
 #define EXTERN_UNLESS_MAIN_MODULE extern
 #else
 #define EXTERN_UNLESS_MAIN_MODULE
diff -ru a/include/cipher.h b/include/cipher.h
--- a/include/cipher.h	2016-04-22 03:30:18.000000000 -0400
+++ b/include/cipher.h	2021-11-17 10:38:42.000000000 -0500
@@ -115,7 +115,7 @@
 typedef struct gcry_md_context *MD_HANDLE;
 
 #ifndef EXTERN_UNLESS_MAIN_MODULE
-#if defined (__riscos__) && !defined (INCLUDED_BY_MAIN_MODULE)
+#if !defined (INCLUDED_BY_MAIN_MODULE)
 #define EXTERN_UNLESS_MAIN_MODULE extern
 #else
 #define EXTERN_UNLESS_MAIN_MODULE
diff -ru a/include/iobuf.h b/include/iobuf.h
--- a/include/iobuf.h	2016-04-22 03:30:18.000000000 -0400
+++ b/include/iobuf.h	2021-11-17 10:38:42.000000000 -0500
@@ -70,7 +70,7 @@
 };
 
 #ifndef EXTERN_UNLESS_MAIN_MODULE
-#if defined (__riscos__) && !defined (INCLUDED_BY_MAIN_MODULE)
+#if !defined (INCLUDED_BY_MAIN_MODULE)
 #define EXTERN_UNLESS_MAIN_MODULE extern
 #else
 #define EXTERN_UNLESS_MAIN_MODULE 
diff -ru a/include/memory.h b/include/memory.h
--- a/include/memory.h	2016-04-22 03:30:18.000000000 -0400
+++ b/include/memory.h	2021-11-17 10:38:42.000000000 -0500
@@ -91,7 +91,7 @@
 #define DBG_MEMSTAT   memory_stat_debug_mode
 
 #ifndef EXTERN_UNLESS_MAIN_MODULE
-#if defined (__riscos__) && !defined (INCLUDED_BY_MAIN_MODULE)
+#if !defined (INCLUDED_BY_MAIN_MODULE)
 #define EXTERN_UNLESS_MAIN_MODULE extern
 #else
 #define EXTERN_UNLESS_MAIN_MODULE 
diff -ru a/include/mpi.h b/include/mpi.h
--- a/include/mpi.h	2016-04-22 03:30:18.000000000 -0400
+++ b/include/mpi.h	2021-11-17 10:38:42.000000000 -0500
@@ -36,7 +36,7 @@
 #include "memory.h"
 
 #ifndef EXTERN_UNLESS_MAIN_MODULE
-#if defined (__riscos__) && !defined (INCLUDED_BY_MAIN_MODULE)
+#if !defined (INCLUDED_BY_MAIN_MODULE)
 #define EXTERN_UNLESS_MAIN_MODULE extern
 #else
 #define EXTERN_UNLESS_MAIN_MODULE 
diff -ru a/keyserver/gpgkeys_curl.c b/keyserver/gpgkeys_curl.c
--- a/keyserver/gpgkeys_curl.c	2016-04-22 03:30:18.000000000 -0400
+++ b/keyserver/gpgkeys_curl.c	2021-11-17 15:34:06.000000000 -0500
@@ -28,6 +28,7 @@
  * wish to do so, delete this exception statement from your version.
  */
 
+#define INCLUDED_BY_MAIN_MODULE 1 /* XXX SmartOS, links to .a library, needed */
 #include <config.h>
 #include <stdio.h>
 #include <string.h>
@@ -42,6 +43,7 @@
 #else
 #include "curl-shim.h"
 #endif
+#include "util.h"
 #include "compat.h"
 #include "keyserver.h"
 #include "ksutil.h"
Only in b/keyserver: gpgkeys_curl.c~
diff -ru a/keyserver/gpgkeys_hkp.c b/keyserver/gpgkeys_hkp.c
--- a/keyserver/gpgkeys_hkp.c	2016-04-22 03:30:18.000000000 -0400
+++ b/keyserver/gpgkeys_hkp.c	2021-11-17 15:34:19.000000000 -0500
@@ -29,6 +29,7 @@
  * wish to do so, delete this exception statement from your version.
  */
 
+#define INCLUDED_BY_MAIN_MODULE 1 /* XXX SmartOS, links to .a library, needed */
 #include <config.h>
 #include <stdio.h>
 #include <string.h>
@@ -60,6 +61,7 @@
 #ifdef USE_DNS_SRV
 # include "srv.h"
 #endif
+#include "util.h"
 #include "compat.h"
 #include "keyserver.h"
 #include "ksutil.h"
Only in b/keyserver: gpgkeys_hkp.c~
diff -ru a/keyserver/gpgkeys_ldap.c b/keyserver/gpgkeys_ldap.c
--- a/keyserver/gpgkeys_ldap.c	2016-04-22 03:30:18.000000000 -0400
+++ b/keyserver/gpgkeys_ldap.c	2021-11-17 15:34:40.000000000 -0500
@@ -29,6 +29,7 @@
  * wish to do so, delete this exception statement from your version.
  */
 
+#define INCLUDED_BY_MAIN_MODULE 1 /* XXX SmartOS, links to .a library, needed */
 #include <config.h>
 #include <stdio.h>
 #include <string.h>
@@ -57,14 +58,11 @@
 #ifdef HAVE_LDAP_SSL_H
 #include <ldap_ssl.h>
 #endif
-
 #include "compat.h"
 #include "keyserver.h"
 #include "ksutil.h"
 
-#ifdef __riscos__
 #include "util.h"
-#endif
 
 #if HAVE_W32_SYSTEM
 # if !defined(__MINGW64_VERSION_MAJOR) || !defined(__MINGW32_MAJOR_VERSION)
Only in b/keyserver: gpgkeys_ldap.c~
diff -ru a/tools/bftest.c b/tools/bftest.c
--- a/tools/bftest.c	2016-04-22 03:30:18.000000000 -0400
+++ b/tools/bftest.c	2021-11-17 11:28:04.000000000 -0500
@@ -26,6 +26,7 @@
 #include <fcntl.h>
 #endif
 
+#define INCLUDED_BY_MAIN_MODULE 1 /* XXX SmartOS, links to .a library, needed */
 #include "util.h"
 #include "cipher.h"
 #include "i18n.h"
diff -ru a/tools/mpicalc.c b/tools/mpicalc.c
--- a/tools/mpicalc.c	2016-04-22 03:30:18.000000000 -0400
+++ b/tools/mpicalc.c	2021-11-17 11:25:42.000000000 -0500
@@ -31,6 +31,7 @@
 #include <stdlib.h>
 #include <ctype.h>
 
+#define INCLUDED_BY_MAIN_MODULE 1 /* XXX SmartOS, links to .a library, needed */
 #include "util.h"
 #include "mpi.h"
 #include "i18n.h"
diff -ru a/tools/shmtest.c b/tools/shmtest.c
--- a/tools/shmtest.c	2016-04-22 03:30:18.000000000 -0400
+++ b/tools/shmtest.c	2021-11-17 10:41:58.000000000 -0500
@@ -25,6 +25,7 @@
 #ifdef HAVE_SYS_SHM_H
 #include <sys/shm.h>
 #endif
+#define INCLUDED_BY_MAIN_MODULE 1 /* XXX SmartOS, links to .a library, needed */
 #include "util.h"
 #include "ttyio.h"
 #include "i18n.h"
