diff -ru a/Makefile.in b/Makefile.in
--- a/Makefile.in	Mon Jun 11 04:47:56 2018
+++ b/Makefile.in	Tue Nov 16 22:44:35 2021
@@ -427,8 +427,7 @@
 # selinux-support is enabled.
 DISTCHECK_CONFIGURE_FLAGS = --enable-mailto
 GITLOG_TO_CHANGELOG = gitlog-to-changelog
-@CROSS_COMPILING_FALSE@checks = checks
-@CROSS_COMPILING_TRUE@checks = 
+checks =
 @HAVE_W32_SYSTEM_FALSE@bzlib = 
 @HAVE_W32_SYSTEM_TRUE@bzlib = bzlib
 SUBDIRS = m4 intl zlib ${bzlib} util mpi cipher tools \
