From d30a346d7376c44fc411cc18764da5649c890ec3 Mon Sep 17 00:00:00 2001
From: Patrick Mooney <pmooney@pfmooney.com>
Date: Wed, 25 Jan 2017 19:32:47 +0000
Subject: [PATCH 4/6] Fix crypto header links

---
 Configure    | 13 +++++++++++++
 Makefile.org |  2 ++
 2 files changed, 15 insertions(+)

diff --git a/Configure b/Configure
index e79678f..537a731 100755
--- a/Configure
+++ b/Configure
@@ -1694,6 +1694,8 @@ unlink("$Makefile.new") || die "unable to remove old $Makefile.new:$!\n" if -e "
 open(OUT,">$Makefile.new") || die "unable to create $Makefile.new:$!\n";
 print OUT "### Generated automatically from Makefile.org by Configure.\n\n";
 my $sdirs=0;
+my $skipdirs=0;
+my @skipped=();
 while (<IN>)
 	{
 	chomp;
@@ -1701,9 +1703,11 @@ while (<IN>)
 	if ($sdirs) {
 		my $dir;
 		foreach $dir (@skip) {
+			push @skipped, $dir if (/$dir\s/ || /$dir$/);
 			s/(\s)$dir /$1/;
 			s/\s$dir$//;
 			}
+		$skipdirs = 1 unless (/\\$/);
 		}
 	$sdirs = 0 unless /\\$/;
         s/engines // if (/^DIRS=/ && $disabled{"engine"});
@@ -1799,6 +1803,15 @@ while (<IN>)
 		}
 	s/^SHARED_LDFLAGS=.*/SHARED_LDFLAGS=$shared_ldflag/;
 	print OUT $_."\n";
+	if ($skipdirs) {
+		my $dir;
+		print OUT "SKIPDIRS= \\\n\t";
+		foreach $dir (@skipped) {
+			print OUT " " . $dir;
+			}
+		print OUT "\n";
+		}
+	$skipdirs = 0;
 	}
 close(IN);
 close(OUT);
diff --git a/Makefile.org b/Makefile.org
index 739b8bb..38f5d32 100644
--- a/Makefile.org
+++ b/Makefile.org
@@ -433,6 +433,8 @@ files:
 	@set -e; target=files; $(RECURSIVE_BUILD_CMD)
 
 links:
+	@$(MAKE) SDIRS='$(SDIRS) $(SKIPDIRS)' all-links
+all-links:
 	@$(PERL) $(TOP)/util/mkdir-p.pl include/openssl
 	@$(PERL) $(TOP)/util/mklink.pl include/openssl $(EXHEADER)
 	@set -e; target=links; $(RECURSIVE_BUILD_CMD)
-- 
2.2.1

