diff --git a/Makefile.org b/Makefile.org
index bc1c3be..580d3e8 100644
--- a/Makefile.org
+++ b/Makefile.org
@@ -170,8 +170,8 @@ ONEDIRS=out tmp
 EDIRS=  times doc bugs util include certs ms shlib mt demos perl sf dep VMS
 WDIRS=  windows
 LIBS=   libcrypto.a libssl.a
-SHARED_CRYPTO=libcrypto$(SHLIB_EXT)
-SHARED_SSL=libssl$(SHLIB_EXT)
+SHARED_CRYPTO=libsunw_crypto$(SHLIB_EXT)
+SHARED_SSL=libsunw_ssl$(SHLIB_EXT)
 SHARED_LIBS=
 SHARED_LIBS_LINK_EXTS=
 SHARED_LDFLAGS=
@@ -180,7 +180,7 @@ GENERAL=        Makefile
 BASENAME=       openssl
 NAME=           $(BASENAME)-$(VERSION)
 TARFILE=        ../$(NAME).tar
-EXHEADER=       e_os2.h
+EXHEADER=       e_os2.h sunw_prefix.h
 HEADER=         e_os.h
 
 all: Makefile build_all
@@ -298,7 +298,7 @@ fips_premain_dso$(EXE_EXT): libcrypto.a
 		$(FIPSLIBDIR)fips_premain.c $(FIPSLIBDIR)fipscanister.o \
 		libcrypto.a $(EX_LIBS)
 
-libcrypto$(SHLIB_EXT): libcrypto.a fips_premain_dso$(EXE_EXT)
+libsunw_crypto$(SHLIB_EXT): libcrypto.a fips_premain_dso$(EXE_EXT)
 	@if [ "$(SHLIB_TARGET)" != "" ]; then \
 		if [ "$(FIPSCANLIB)" = "libcrypto" ]; then \
 			FIPSLD_LIBCRYPTO=libcrypto.a ; \
@@ -312,9 +312,9 @@ libcrypto$(SHLIB_EXT): libcrypto.a fips_premain_dso$(EXE_EXT)
 		exit 1; \
 	fi
 
-libssl$(SHLIB_EXT): libcrypto$(SHLIB_EXT) libssl.a
+libsunw_ssl$(SHLIB_EXT): libsunw_crypto$(SHLIB_EXT) libssl.a
 	@if [ "$(SHLIB_TARGET)" != "" ]; then \
-		$(MAKE) SHLIBDIRS=ssl SHLIBDEPS='-lcrypto' build-shared; \
+		$(MAKE) SHLIBDIRS=ssl SHLIBDEPS='-lsunw_crypto' build-shared; \
 	else \
 		echo "There's no support for shared libraries on this platform" >&2; \
 		exit 1; \
@@ -340,7 +340,7 @@ link-shared:
 			LIBNAME=$$i LIBVERSION=$(SHLIB_MAJOR).$(SHLIB_MINOR) \
 			LIBCOMPATVERSIONS=";$(SHLIB_VERSION_HISTORY)" \
 			symlink.$(SHLIB_TARGET); \
-		libs="$$libs -l$$i"; \
+		libs="$$libs -lsunw_$$i"; \
 	done
 
 build-shared: do_$(SHLIB_TARGET) link-shared
@@ -355,7 +355,7 @@ do_$(SHLIB_TARGET):
 			LIBCOMPATVERSIONS=";$(SHLIB_VERSION_HISTORY)" \
 			LIBDEPS="$$libs $(EX_LIBS)" \
 			link_a.$(SHLIB_TARGET); \
-		libs="-l$$i $$libs"; \
+		libs="-lsunw_$$i $$libs"; \
 	done
 
 libcrypto.pc: Makefile
diff --git a/Makefile.shared b/Makefile.shared
index e753f44..b1a79d4 100644
--- a/Makefile.shared
+++ b/Makefile.shared
@@ -368,7 +368,7 @@ link_o.solaris:
 		$(CALC_VERSIONS); \
 		MINUSZ='-z '; \
 		($(CC) -v 2>&1 | grep gcc) > /dev/null && MINUSZ='-Wl,-z,'; \
-		SHLIB=lib$(LIBNAME).so; \
+		SHLIB=libsunw_$(LIBNAME).so; \
 		SHLIB_SUFFIX=; \
 		ALLSYMSFLAGS="$${MINUSZ}allextract"; \
 		NOALLSYMSFLAGS="$${MINUSZ}defaultextract"; \
@@ -382,7 +382,7 @@ link_a.solaris:
 		$(CALC_VERSIONS); \
 		MINUSZ='-z '; \
 		($(CC) -v 2>&1 | grep gcc) > /dev/null && MINUSZ='-Wl,-z,'; \
-		SHLIB=lib$(LIBNAME).so; \
+		SHLIB=libsunw_$(LIBNAME).so; \
 		SHLIB_SUFFIX=;\
 		ALLSYMSFLAGS="$${MINUSZ}allextract"; \
 		NOALLSYMSFLAGS="$${MINUSZ}defaultextract"; \
diff --git a/apps/Makefile b/apps/Makefile
index 8c3297e..0477578 100644
--- a/apps/Makefile
+++ b/apps/Makefile
@@ -26,8 +26,8 @@ GENERAL=Makefile makeapps.com install.com
 
 DLIBCRYPTO=../libcrypto.a
 DLIBSSL=../libssl.a
-LIBCRYPTO=-L.. -lcrypto
-LIBSSL=-L.. -lssl
+LIBCRYPTO=-L.. -lsunw_crypto
+LIBSSL=-L.. -lsunw_ssl
 
 PROGRAM= openssl
 
diff --git a/crypto/Makefile b/crypto/Makefile
index 618c958..ee087f8 100644
--- a/crypto/Makefile
+++ b/crypto/Makefile
@@ -34,7 +34,7 @@ GENERAL=Makefile README crypto-lib.com install.com
 TEST=constant_time_test.c
 
 LIB= $(TOP)/libcrypto.a
-SHARED_LIB= libcrypto$(SHLIB_EXT)
+SHARED_LIB= libsunw_crypto$(SHLIB_EXT)
 LIBSRC=	cryptlib.c mem.c mem_clr.c mem_dbg.c cversion.c ex_data.c cpt_err.c \
 	ebcdic.c uid.c o_time.c o_str.c o_dir.c o_fips.c o_init.c fips_ers.c
 LIBOBJ= cryptlib.o mem.o mem_dbg.o cversion.o ex_data.o cpt_err.o ebcdic.o \
diff --git a/crypto/opensslconf.h.in b/crypto/opensslconf.h.in
index 67e5cff..5e8ae4d 100644
--- a/crypto/opensslconf.h.in
+++ b/crypto/opensslconf.h.in
@@ -4,6 +4,8 @@
 #error	"SPARC is not currently supported by the build system."
 #endif
 
+#include <openssl/sunw_prefix.h>
+
 /* Generate 80386 code? */
 #undef I386_ONLY
 
diff --git a/crypto/perlasm/x86_64-xlate.pl b/crypto/perlasm/x86_64-xlate.pl
index b262b8d..cee9daf 100755
--- a/crypto/perlasm/x86_64-xlate.pl
+++ b/crypto/perlasm/x86_64-xlate.pl
@@ -68,7 +68,7 @@ open STDOUT,">$output" || die "can't open $output: $!"
 my $gas=1;	$gas=0 if ($output =~ /\.asm$/);
 my $elf=1;	$elf=0 if (!$gas);
 my $win64=0;
-my $prefix="";
+my $prefix="sunw_";
 my $decor=".L";
 
 my $masmref=8 + 50727*2**-32;	# 8.00.50727 shipped with VS2005
@@ -467,6 +467,7 @@ my %globals;
 				    }
 				    $line =~ s/\@abi\-omnipotent/\@function/;
 				    $line =~ s/\@function.*/\@function/;
+				    $line =~ s/$sym/$globals{$sym} or $sym/e;
 				    last;
 				  };
 		/\.asciz/   && do { if ($line =~ /^"(.*)"$/) {
@@ -480,6 +481,9 @@ my %globals;
 				    $line =~ s/\.L/$decor/g;
 				    last;
 				  };
+		/\.size/    && do { $line =~ s/([_a-z][_a-z0-9]*)/$globals{$1} or $1/gei;
+				    last;
+				  };
 	    }
 
 	    if ($gas) {
@@ -513,6 +517,7 @@ my %globals;
 		} elsif ($dir =~ /\.hidden/) {
 		    if    ($flavour eq "macosx")  { $self->{value} = ".private_extern\t$prefix$line"; }
 		    elsif ($flavour eq "mingw64") { $self->{value} = ""; }
+		    else { $self->{value} = ".hidden\t$prefix$line"; }
 		} elsif ($dir =~ /\.comm/) {
 		    $self->{value} = "$dir\t$prefix$line";
 		    $self->{value} =~ s|,([0-9]+),([0-9]+)$|",$1,".log($2)/log(2)|e if ($flavour eq "macosx");
diff --git a/crypto/perlasm/x86asm.pl b/crypto/perlasm/x86asm.pl
index eb543db..102dadf 100644
--- a/crypto/perlasm/x86asm.pl
+++ b/crypto/perlasm/x86asm.pl
@@ -137,6 +137,7 @@ $label="000";
 
 sub ::islabel		# see is argument is a known label
 { my $i;
+    if ($_[0] eq "_GLOBAL_OFFSET_TABLE_") { return $_[0]; }
     foreach $i (values %label) { return $i if ($i eq $_[0]); }
   $label{$_[0]};	# can be undef
 }
diff --git a/crypto/perlasm/x86gas.pl b/crypto/perlasm/x86gas.pl
index 682a3a3..86708ec 100644
--- a/crypto/perlasm/x86gas.pl
+++ b/crypto/perlasm/x86gas.pl
@@ -5,7 +5,7 @@ package x86gas;
 *out=\@::out;
 
 $::lbdecor=$::aout?"L":".L";		# local label decoration
-$nmdecor=($::aout or $::coff)?"_":"";	# external name decoration
+$nmdecor=($::aout or $::coff)?"_":"sunw_";	# external name decoration
 
 $initseg="";
 
diff --git a/ssl/Makefile b/ssl/Makefile
index f12d69a..d3707fe 100644
--- a/ssl/Makefile
+++ b/ssl/Makefile
@@ -19,7 +19,7 @@ TEST=ssltest.c heartbeat_test.c clienthellotest.c
 APPS=
 
 LIB=$(TOP)/libssl.a
-SHARED_LIB= libssl$(SHLIB_EXT)
+SHARED_LIB= libsunw_ssl$(SHLIB_EXT)
 LIBSRC=	\
 	s2_meth.c   s2_srvr.c s2_clnt.c  s2_lib.c  s2_enc.c s2_pkt.c \
 	s3_meth.c   s3_srvr.c s3_clnt.c  s3_lib.c  s3_enc.c s3_pkt.c s3_both.c s3_cbc.c \
