diff -ur ipmitool-1.8.11/configure ipmitool-1.8.11.n/configure
--- ipmitool-1.8.11/configure	2009-02-25 21:21:56.000000000 +0000
+++ ipmitool-1.8.11.n/configure	2012-06-20 16:17:48.553434839 +0000
@@ -22080,8 +22080,6 @@
 	STRIP=gstrip
 	LD=gld
 	AR=gar
-	# openssl libs are in /usr/sfw/lib on solaris 10
-        LIBS="$LIBS -R/usr/sfw/lib"
 	# disable the linux-specific interfaces
 	# and bmc on SPARC (there is no bmc driver for SPARC)
 	if [ `mach -p` = sparc ]; then
diff -ur ipmitool-1.8.11/configure.in ipmitool-1.8.11.n/configure.in
--- ipmitool-1.8.11/configure.in	2012-04-23 16:01:34.000000000 +0000
+++ ipmitool-1.8.11.n/configure.in	2012-06-20 16:17:55.126213525 +0000
@@ -60,8 +60,6 @@
 	STRIP=gstrip
 	LD=gld
 	AR=gar
-	# openssl libs are in /usr/sfw/lib on solaris 10
-        LIBS="$LIBS -R/usr/sfw/lib"
 	# disable the linux-specific interfaces
 	enable_intf_bmc=no
 	enable_intf_imb=no
