From 0fd1de72fd63d40007023caf9956de49f99de03a Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Wed, 5 Aug 2015 12:25:15 -0700
Subject: [PATCH 21/34] Try to create privsep chroot dir if it doesn't exist
 yet

---
 sshd.c | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/sshd.c b/sshd.c
index 994b7c9e..e09f3934 100644
--- a/sshd.c
+++ b/sshd.c
@@ -1870,11 +1870,30 @@ main(int ac, char **av)
 
 	if (privsep_chroot) {
 		struct stat st;
+		int rc;
 
 		if ((stat(_PATH_PRIVSEP_CHROOT_DIR, &st) == -1) ||
-		    (S_ISDIR(st.st_mode) == 0))
-			fatal("Missing privilege separation directory: %s",
-			    _PATH_PRIVSEP_CHROOT_DIR);
+		    (S_ISDIR(st.st_mode) == 0)) {
+			rc = mkdir(_PATH_PRIVSEP_CHROOT_DIR, 0755);
+			if (rc == 0) {
+				/*
+				 * If mkdir works, try stat again, so the
+				 * permissions check below can work.
+				 */
+				rc = stat(_PATH_PRIVSEP_CHROOT_DIR, &st);
+				if (rc == 0 && S_ISDIR(st.st_mode) == 0) {
+					rc = -1;
+					errno = ENOTDIR;
+				}
+			}
+
+			if (rc != 0) {
+				fatal("Failed to create privilege separation "
+				    "directory %s: %s",
+				    _PATH_PRIVSEP_CHROOT_DIR,
+				    strerror(errno));
+			}
+		}
 
 #ifdef HAVE_CYGWIN
 		if (check_ntsec(_PATH_PRIVSEP_CHROOT_DIR) &&
-- 
2.22.0

