From 28163c3c880644eb07e8a7f97f1c1b1c43d8f981 Mon Sep 17 00:00:00 2001
From: Dan McDonald <danmcd@mnx.io>
Date: Wed, 7 May 2025 10:21:00 -0400
Subject: [PATCH 32/34] gcc14 fixes for illumos

---
 auth-pam.c | 6 +++++-
 scp.c      | 4 ++--
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/auth-pam.c b/auth-pam.c
index 22a77bab4..b713e209c 100644
--- a/auth-pam.c
+++ b/auth-pam.c
@@ -69,7 +69,11 @@
 
 /* OpenGroup RFC86.0 and XSSO specify no "const" on arguments */
 #ifdef PAM_SUN_CODEBASE
-# define sshpam_const		/* Solaris, HP-UX, SunOS */
+# ifdef __illumos__
+#   define sshpam_const const	/* illumos */
+# else
+#   define sshpam_const		/* Solaris, HP-UX, SunOS */
+# endif
 #else
 # define sshpam_const	const	/* LinuxPAM, OpenPAM, AIX */
 #endif
diff --git a/scp.c b/scp.c
index 57c242ffc..2b9d6821f 100644
--- a/scp.c
+++ b/scp.c
@@ -207,7 +207,7 @@ killchild(int signo)
 }
 
 static void
-suspone(int pid, int signo)
+suspone(pid_t pid, int signo)
 {
 	int status;
 
@@ -1017,7 +1017,7 @@ brace_expand(const char *pattern, char ***patternsp, size_t *npatternsp)
 
 static struct sftp_conn *
 do_sftp_connect(char *host, char *user, int port, char *sftp_direct,
-   int *reminp, int *remoutp, int *pidp)
+   int *reminp, int *remoutp, pid_t *pidp)
 {
 	if (sftp_direct == NULL) {
 		if (do_cmd(ssh_program, host, user, port, 1, "sftp",
-- 
2.48.1

