diff -ruN a/shell.c b/shell.c
--- a/shell.c	2022-03-04 22:13:00.000000000 +0000
+++ b/shell.c	2024-11-21 14:58:02.408512324 +0000
@@ -236,7 +236,7 @@
 #else
 int posixly_correct = 0;	/* Non-zero means posix.2 superset. */
 #endif
-
+int import_functions = 0;	/* Import functions from environment */
 /* Some long-winded argument names.  These are obviously new. */
 #define Int 1
 #define Charp 2
@@ -257,6 +257,7 @@
   { "help", Int, &want_initial_help, (char **)0x0 },
   { "init-file", Charp, (int *)0x0, &bashrc_file },
   { "login", Int, &make_login_shell, (char **)0x0 },
+  { "import-functions", Int, &import_functions, (char **)0x0 },
   { "noediting", Int, &no_line_editing, (char **)0x0 },
   { "noprofile", Int, &no_profile, (char **)0x0 },
   { "norc", Int, &no_rc, (char **)0x0 },
diff -ruN a/variables.c b/variables.c
--- a/variables.c	2022-07-18 21:31:35.000000000 +0000
+++ b/variables.c	2024-11-21 14:58:46.140981394 +0000
@@ -113,6 +113,7 @@
 /* Variables used here and defined in other files. */
 extern time_t shell_start_time;
 extern struct timeval shellstart;
+extern int import_functions;
 
 /* The list of shell variables that the user has created at the global
    scope, or that came from the environment. */
@@ -394,7 +395,7 @@
 #if defined (FUNCTION_IMPORT)
       /* If exported function, define it now.  Don't import functions from
 	 the environment in privileged mode. */
-      if (privmode == 0 && read_but_dont_execute == 0 && 
+      if (import_functions && privmode == 0 && read_but_dont_execute == 0 && 
           STREQN (BASHFUNC_PREFIX, name, BASHFUNC_PREFLEN) &&
           STREQ (BASHFUNC_SUFFIX, name + char_index - BASHFUNC_SUFFLEN) &&
 	  STREQN ("() {", string, 4))
