Common subdirectories: a/builtins and b/builtins
diff -ru a/configure b/configure
--- a/configure	Thu Nov 21 12:41:03 2024
+++ b/configure	Thu Nov 21 12:58:04 2024
@@ -5284,10 +5284,12 @@
 /* end confdefs.h.  */
 
 #include <stdio.h>
+#include <stdlib.h>
 #include <readline/readline.h>
 
 extern int rl_gnu_readline_p;
 
+int
 main()
 {
 	FILE *fp;
@@ -11264,6 +11266,7 @@
 #include <locale.h>
 #include <wchar.h>
 
+int
 main(c, v)
 int     c;
 char    **v;
@@ -12662,9 +12665,11 @@
 #ifdef HAVE_LOCALE_H
 #include <locale.h>
 #endif
+#include <stdlib.h>
 #include <stdio.h>
 #include <ctype.h>
 
+int
 main(c, v)
 int	c;
 char	*v[];
@@ -12723,6 +12728,8 @@
 
 #include <sys/types.h>
 #include <fcntl.h>
+#include <stdlib.h>
+int
 main()
 {
   int fd1, fd2, fl;
@@ -12774,6 +12781,8 @@
 #ifdef HAVE_UNISTD_H
 #  include <unistd.h>
 #endif
+#include <stdlib.h>
+int
 main()
 {
 # ifdef GETPGRP_VOID
@@ -12984,6 +12993,7 @@
 
 #include <sys/types.h>
 #include <signal.h>
+#include <stdlib.h>
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif
@@ -12990,6 +13000,7 @@
 #if !HAVE_DECL_SYS_SIGLIST
 extern char *sys_siglist[];
 #endif
+int
 main()
 {
 char *msg = sys_siglist[2];
@@ -13071,6 +13082,8 @@
 #ifndef UNDER_SYS_SIGLIST_DECLARED
 extern char *_sys_siglist[];
 #endif
+#include <stdlib.h>
+int
 main()
 {
 char *msg = (char *)_sys_siglist[2];
@@ -13475,6 +13488,8 @@
 #include <sys/types.h>
 #include <sys/time.h>
 #include <sys/resource.h>
+#include <stdlib.h>
+int
 main()
 {
 #ifdef HAVE_QUAD_T
@@ -14002,6 +14017,7 @@
 
 #include <sys/wait.h>
 
+int
 main(c, v)
      int c;
      char **v;
@@ -14314,6 +14330,7 @@
 /* end confdefs.h.  */
 
 #include <stdio.h>
+#include <stdlib.h>
 #include <sys/types.h>
 #include <fcntl.h>
 #ifdef HAVE_UNISTD_H
@@ -14333,6 +14350,7 @@
 #  include <ndir.h>
 # endif
 #endif /* HAVE_DIRENT_H */
+int
 main()
 {
 DIR *dir;
@@ -14384,6 +14402,8 @@
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
+#include <stdlib.h>
+int
 main()
 {
 long maxfds = ulimit(4, 0L);
@@ -14461,6 +14481,8 @@
 {
 return "42";
 }
+#include <stdlib.h>
+int
 main()
 {
 char *s;
@@ -14512,7 +14534,9 @@
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif
+#include <stdlib.h>
 
+int
 main()
 {
 	char	*xpwd;
@@ -14567,7 +14591,9 @@
 #include <sys/types.h>
 #include <signal.h>
 #include <setjmp.h>
+#include <stdlib.h>
 
+int
 main()
 {
 #if !defined (_POSIX_VERSION) || !defined (HAVE_POSIX_SIGNALS)
@@ -14636,7 +14662,9 @@
 #if defined (HAVE_LOCALE_H)
 #include <locale.h>
 #endif
+#include <stdlib.h>
 
+int
 main(c, v)
 int     c;
 char    *v[];
@@ -14706,7 +14734,9 @@
 /* end confdefs.h.  */
 
 #include <stdio.h>
+#include <stdlib.h>
 
+int
 main()
 {
   int n;
@@ -14785,7 +14815,9 @@
   va_end (args);
   return n;
 }
+#include <stdlib.h>
 
+int
 main()
 {
   int n;
@@ -14945,6 +14977,7 @@
 
 #include <stdio.h>
 #include <string.h>
+#include <stdlib.h>
 
 int
 main()
@@ -14991,6 +15024,7 @@
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
+#include <stdlib.h>
 
 #include <signal.h>
 #ifdef HAVE_UNISTD_H
@@ -15026,6 +15060,7 @@
   nsigint++;
 }
 
+int
 main()
 {
 	nsigint = 0;
@@ -15077,8 +15112,10 @@
 #include <unistd.h>
 #endif
 #include <signal.h>
+#include <stdlib.h>
 
 /* Add more tests in here as appropriate. */
+int
 main()
 {
 /* signal type */
@@ -15150,8 +15187,10 @@
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif
+#include <stdlib.h>
 
 /* Add more tests in here as appropriate. */
+int
 main()
 {
 int fd, err;
@@ -15350,6 +15389,8 @@
 #ifndef errno
 extern int errno;
 #endif
+#include <stdlib.h>
+int
 main()
 {
 	int	x;
@@ -15464,7 +15505,9 @@
 #ifndef NSIG
 #  define NSIG 64
 #endif
+#include <stdlib.h>
 
+int
 main ()
 {
   int n_sigs = 2 * NSIG;
diff -ru a/execute_cmd.c b/execute_cmd.c
--- a/execute_cmd.c	Thu Nov 21 13:09:19 2024
+++ b/execute_cmd.c	Thu Nov 21 13:09:48 2024
@@ -2061,6 +2061,7 @@
 void
 coproc_pidchk (pid, status)
      pid_t pid;
+     int status; /* XXX SmartOS - gcc14 requirement */
 {
   struct coproc *cp;
 
Common subdirectories: a/lib and b/lib
diff -ru a/builtins/command.def b/builtins/command.def
--- a/builtins/command.def	Thu Nov 21 13:16:59 2024
+++ b/builtins/command.def	Thu Nov 21 13:23:50 2024
@@ -59,6 +59,8 @@
 #if defined (_CS_PATH) && defined (HAVE_CONFSTR) && !HAVE_DECL_CONFSTR
 extern size_t confstr __P((int, char *, size_t));
 #endif
+/* XXX SmartOS - confstr isn't defined in spite of above!!! Fix! */
+extern size_t confstr(int, char *, size_t);
 
 extern int subshell_environment;
 
Common subdirectories: a/lib/readline and b/lib/readline
diff -ru a/lib/readline/tcap.h b/lib/readline/tcap.h
--- a/lib/readline/tcap.h	Thu Nov 21 13:31:02 2024
+++ b/lib/readline/tcap.h	Thu Nov 21 13:34:35 2024
@@ -30,7 +30,7 @@
 #  if defined (__linux__) && !defined (SPEED_T_IN_SYS_TYPES)
 #    include "rltty.h"
 #  endif
-#  include <termcap.h>
+#  include <../xpg4/include/termcap.h> /* XXX SmartOS, use xpg4!!! */
 #else
 
 /* On Solaris2, sys/types.h #includes sys/reg.h, which #defines PC.
