From 2264bfb2d1960e2d571c05a6798b66396f2aee0a Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Sat, 18 Jan 2014 07:48:55 +0000
Subject: [PATCH] runpath.patch

---
 configure    | 29 -----------------------------
 configure.in | 29 -----------------------------
 2 files changed, 58 deletions(-)

diff --git a/configure b/configure
index 2796526..6e9e665 100755
--- a/configure
+++ b/configure
@@ -26776,35 +26776,6 @@ else
 fi
 
 NEWFLAGS=""
-for e in $BUILD_LDFLAGS ; do
-    case $e in
-	-L*)
-	    case $host_os in
-		netbsd*)
-		    ee=`echo $e | sed -e 's%^-L%-Wl,-rpath,%'`
-		    NEWFLAGS="$NEWFLAGS $e $ee"
-		    ;;
-		freebsd*)
-		    ee=`echo $e | sed -e 's%^-L%-Wl,-rpath,%'`
-		    NEWFLAGS="$NEWFLAGS $e $ee"
-		    ;;
-		solaris*)
-		    ee=`echo $e | sed -e 's%^-L%-R%'`
-		    NEWFLAGS="$NEWFLAGS $e $ee"
-		    ;;
-		*)
-		    NEWFLAGS="$NEWFLAGS $e"
-		    ;;
-		esac
-	    ;;
-	*)
-	    NEWFLAGS="$NEWFLAGS $e"
-	    ;;
-    esac
-done
-BUILD_LDFLAGS="$NEWFLAGS"
-
-NEWFLAGS=""
 for e in $DNS_GSSAPI_LIBS ; do
     case $e in
 	-L*)
diff --git a/configure.in b/configure.in
index d5fb15c..b9fdd70 100644
--- a/configure.in
+++ b/configure.in
@@ -3165,35 +3165,6 @@ else
 fi
 
 NEWFLAGS=""
-for e in $BUILD_LDFLAGS ; do
-    case $e in
-	-L*)
-	    case $host_os in
-		netbsd*)
-		    ee=`echo $e | sed -e 's%^-L%-Wl,-rpath,%'`
-		    NEWFLAGS="$NEWFLAGS $e $ee"
-		    ;;
-		freebsd*)
-		    ee=`echo $e | sed -e 's%^-L%-Wl,-rpath,%'`
-		    NEWFLAGS="$NEWFLAGS $e $ee"
-		    ;;
-		solaris*)
-		    ee=`echo $e | sed -e 's%^-L%-R%'`
-		    NEWFLAGS="$NEWFLAGS $e $ee"
-		    ;;
-		*)
-		    NEWFLAGS="$NEWFLAGS $e"
-		    ;;
-		esac
-	    ;;
-	*)
-	    NEWFLAGS="$NEWFLAGS $e"
-	    ;;
-    esac
-done
-BUILD_LDFLAGS="$NEWFLAGS"
-
-NEWFLAGS=""
 for e in $DNS_GSSAPI_LIBS ; do
     case $e in
 	-L*)
-- 
1.8.4

