--- a/nss/lib/jar/jarfile.c	2016-06-20 17:11:28.000000000 +0000
+++ b/nss/lib/jar/jarfile.c	2024-10-31 11:36:22.718304546 +0000
@@ -657,7 +657,7 @@
 jar_listzip(JAR *jar, JAR_FILE fp)
 {
     ZZLink *ent;
-    JAR_Item *it;
+    JAR_Item *it = NULL;
     JAR_Physical *phy = NULL;
     struct ZipLocal *Local = PORT_ZNew(struct ZipLocal);
     struct ZipCentral *Central = PORT_ZNew(struct ZipCentral);
