--- ignore/nss/lib/sqlite/Makefile.orig	Tue Sep  6 23:47:39 2016
+++ ignore/nss/lib/sqlite/Makefile	Tue Sep  6 23:49:28 2016
@@ -51,3 +51,7 @@
 OS_CFLAGS += -w44996
 endif
 
+ifeq (SunOS,$(OS_ARCH))
+# sqlite needs C99 for access to newer SUS standards
+OS_CFLAGS += -std=gnu99
+endif
--- a/nss/lib/sqlite/sqlite3.c	2016-06-20 13:11:28.000000000 -0400
+++ b/nss/lib/sqlite/sqlite3.c	2021-11-15 23:21:35.000000000 -0500
@@ -109780,8 +109780,12 @@
   pNew = sqlite3DbMallocZero(db, sizeof(*pNew) );
   if( pNew==0 ){
     assert( db->mallocFailed );
+#if 0
     pNew = &standin;
     memset(pNew, 0, sizeof(*pNew));
+#else
+    return NULL;
+#endif
   }
   if( pEList==0 ){
     pEList = sqlite3ExprListAppend(pParse, 0, sqlite3Expr(db,TK_ASTERISK,0));
