diff -urN node-v0.6.8.orig/deps/uv/src/unix/pipe.c node-v0.6.8/deps/uv/src/unix/pipe.c
--- node-v0.6.8.orig/deps/uv/src/unix/pipe.c	2012-01-23 18:29:41.370902485 +0000
+++ node-v0.6.8/deps/uv/src/unix/pipe.c	2012-01-23 18:39:09.747298075 +0000
@@ -251,7 +251,6 @@
   pipe = watcher->data;
 
   assert(pipe->type == UV_NAMED_PIPE);
-  assert(pipe->pipe_fname != NULL);
 
   sockfd = uv__accept(pipe->fd, (struct sockaddr *)&saddr, sizeof saddr);
   if (sockfd == -1) {
