diff -ru b/configure c/configure
--- b/configure	Thu Oct  5 17:10:35 2023
+++ c/configure	Thu Oct  5 17:17:48 2023
@@ -30166,7 +30166,7 @@
          $PKGCONFIG --cflags-only-I libidn2 2>/dev/null`
       IDN_DIR=`echo $IDN_LDFLAGS | $SED -e 's/^-L//'`
     else
-            IDN_LIBS="-lidn2"
+            IDN_LIBS="-ljoy_idn2"
       IDN_LDFLAGS="-L$want_idn_path/lib$libsuff"
       IDN_CPPFLAGS="-I$want_idn_path/include"
       IDN_DIR="$want_idn_path/lib$libsuff"
@@ -30314,7 +30314,7 @@
       IDN_CPPFLAGS=`$PKGCONFIG --cflags-only-I libidn2 2>/dev/null`
       IDN_DIR=`echo $IDN_LDFLAGS | $SED -e 's/^-L//'`
     else
-            IDN_LIBS="-lidn2"
+            IDN_LIBS="-ljoy_idn2"
     fi
   fi
   #
diff -ru b/configure.ac c/configure.ac
--- b/configure.ac	Wed Sep 13 02:18:47 2023
+++ c/configure.ac	Thu Oct  5 17:18:04 2023
@@ -2674,7 +2674,7 @@
       IDN_DIR=`echo $IDN_LDFLAGS | $SED -e 's/^-L//'`
     else
       dnl pkg-config not available or provides no info
-      IDN_LIBS="-lidn2"
+      IDN_LIBS="-ljoy_idn2"
       IDN_LDFLAGS="-L$want_idn_path/lib$libsuff"
       IDN_CPPFLAGS="-I$want_idn_path/include"
       IDN_DIR="$want_idn_path/lib$libsuff"
@@ -2689,7 +2689,7 @@
       IDN_DIR=`echo $IDN_LDFLAGS | $SED -e 's/^-L//'`
     else
       dnl pkg-config not available or provides no info
-      IDN_LIBS="-lidn2"
+      IDN_LIBS="-ljoy_idn2"
     fi
   fi
   #
