diff -ru a/configure b/configure
--- a/configure	Wed Sep 11 01:37:35 2024
+++ b/configure	Wed Sep 11 18:29:49 2024
@@ -8416,7 +8416,7 @@
     We can't simply define LARGE_OFF_T to be 9223372036854775807,
     since some C++ compilers masquerading as C compilers
     incorrectly reject 9223372036854775807.  */
-#define LARGE_OFF_T (((off_t) 1 << 31 << 31) - 1 + ((off_t) 1 << 31 << 31))
+#define LARGE_OFF_T (((off_t) 1ULL << 62) - 1 + ((off_t) 1ULL << 62))
   int off_t_is_large[(LARGE_OFF_T % 2147483629 == 721
 		       && LARGE_OFF_T % 2147483647 == 1)
 		      ? 1 : -1];
@@ -8464,7 +8464,7 @@
     We can't simply define LARGE_OFF_T to be 9223372036854775807,
     since some C++ compilers masquerading as C compilers
     incorrectly reject 9223372036854775807.  */
-#define LARGE_OFF_T (((off_t) 1 << 31 << 31) - 1 + ((off_t) 1 << 31 << 31))
+#define LARGE_OFF_T (((off_t) 1ULL << 62) - 1 + ((off_t) 1ULL << 62))
   int off_t_is_large[(LARGE_OFF_T % 2147483629 == 721
 		       && LARGE_OFF_T % 2147483647 == 1)
 		      ? 1 : -1];
@@ -8488,7 +8488,7 @@
     We can't simply define LARGE_OFF_T to be 9223372036854775807,
     since some C++ compilers masquerading as C compilers
     incorrectly reject 9223372036854775807.  */
-#define LARGE_OFF_T (((off_t) 1 << 31 << 31) - 1 + ((off_t) 1 << 31 << 31))
+#define LARGE_OFF_T (((off_t) 1ULL << 62) - 1 + ((off_t) 1ULL << 62))
   int off_t_is_large[(LARGE_OFF_T % 2147483629 == 721
 		       && LARGE_OFF_T % 2147483647 == 1)
 		      ? 1 : -1];
@@ -8532,7 +8532,7 @@
     We can't simply define LARGE_OFF_T to be 9223372036854775807,
     since some C++ compilers masquerading as C compilers
     incorrectly reject 9223372036854775807.  */
-#define LARGE_OFF_T (((off_t) 1 << 31 << 31) - 1 + ((off_t) 1 << 31 << 31))
+#define LARGE_OFF_T (((off_t) 1ULL << 62) - 1 + ((off_t) 1ULL << 62))
   int off_t_is_large[(LARGE_OFF_T % 2147483629 == 721
 		       && LARGE_OFF_T % 2147483647 == 1)
 		      ? 1 : -1];
@@ -8556,7 +8556,7 @@
     We can't simply define LARGE_OFF_T to be 9223372036854775807,
     since some C++ compilers masquerading as C compilers
     incorrectly reject 9223372036854775807.  */
-#define LARGE_OFF_T (((off_t) 1 << 31 << 31) - 1 + ((off_t) 1 << 31 << 31))
+#define LARGE_OFF_T (((off_t) 1ULL << 62) - 1 + ((off_t) 1ULL << 62))
   int off_t_is_large[(LARGE_OFF_T % 2147483629 == 721
 		       && LARGE_OFF_T % 2147483647 == 1)
 		      ? 1 : -1];
@@ -24828,7 +24828,7 @@
     '-llber -lldap' \
     '-lldapssl -lldapx -lldapsdk' \
     '-lldapsdk -lldapx -lldapssl' \
-    '-lldap -llber -lssl -lcrypto' ; do
+    '-lldap -llber -lsunw_ssl -lsunw_crypto' ; do
 
     if test "$curl_cv_ldap_LIBS" = "unknown"; then
       if test -z "$x_nlibs"; then
@@ -26567,14 +26567,14 @@
     CPPFLAGS="$CPPFLAGS $SSL_CPPFLAGS"
   LDFLAGS="$LDFLAGS $SSL_LDFLAGS"
 
-  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for HMAC_Update in -lcrypto" >&5
-printf %s "checking for HMAC_Update in -lcrypto... " >&6; }
-if test ${ac_cv_lib_crypto_HMAC_Update+y}
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for sunw_HMAC_Update in -lsunw_crypto" >&5
+printf %s "checking for sunw_HMAC_Update in -lsunw_crypto... " >&6; }
+if test ${ac_cv_lib_sunw_crypto_sunw_HMAC_Update+y}
 then :
   printf %s "(cached) " >&6
 else $as_nop
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto  $LIBS"
+LIBS="-lsunw_crypto  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -26582,10 +26582,10 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char HMAC_Update ();
+char sunw_HMAC_Update ();
 int main (void)
 {
-return HMAC_Update ();
+return sunw_HMAC_Update ();
  ;
  return 0;
 }
@@ -26592,21 +26592,21 @@
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
 then :
-  ac_cv_lib_crypto_HMAC_Update=yes
+  ac_cv_lib_crypto_sunw_HMAC_Update=yes
 else $as_nop
-  ac_cv_lib_crypto_HMAC_Update=no
+  ac_cv_lib_crypto_sunw_HMAC_Update=no
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.beam \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_HMAC_Update" >&5
-printf "%s\n" "$ac_cv_lib_crypto_HMAC_Update" >&6; }
-if test "x$ac_cv_lib_crypto_HMAC_Update" = xyes
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_HMAC_Update" >&5
+printf "%s\n" "$ac_cv_lib_sunw_crypto_sunw_HMAC_Update" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_HMAC_Update" = xyes
 then :
 
     HAVECRYPTO="yes"
-    LIBS="-lcrypto $LIBS"
+    LIBS="-lsunw_crypto $LIBS"
 
 else $as_nop
 
@@ -26620,14 +26620,14 @@
     # Linking previously failed, try extra paths from --with-openssl or
     # pkg-config.  Use a different function name to avoid reusing the earlier
     # cached result.
-    { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for HMAC_Init_ex in -lcrypto" >&5
-printf %s "checking for HMAC_Init_ex in -lcrypto... " >&6; }
-if test ${ac_cv_lib_crypto_HMAC_Init_ex+y}
+    { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for sunw_HMAC_Init_ex in -lsunw_crypto" >&5
+printf %s "checking for sunw_HMAC_Init_ex in -lsunw_crypto... " >&6; }
+if test ${ac_cv_lib_sunw_crypto_sunw_HMAC_Init_ex+y}
 then :
   printf %s "(cached) " >&6
 else $as_nop
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto  $LIBS"
+LIBS="-lsunw_crypto  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -26635,10 +26635,10 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char HMAC_Init_ex ();
+char sunw_HMAC_Init_ex ();
 int main (void)
 {
-return HMAC_Init_ex ();
+return sunw_HMAC_Init_ex ();
  ;
  return 0;
 }
@@ -26645,27 +26645,27 @@
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
 then :
-  ac_cv_lib_crypto_HMAC_Init_ex=yes
+  ac_cv_lib_sunw_crypto_sunw_HMAC_Init_ex=yes
 else $as_nop
-  ac_cv_lib_crypto_HMAC_Init_ex=no
+  ac_cv_lib_sunw_crypto_sunw_HMAC_Init_ex=no
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.beam \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_HMAC_Init_ex" >&5
-printf "%s\n" "$ac_cv_lib_crypto_HMAC_Init_ex" >&6; }
-if test "x$ac_cv_lib_crypto_HMAC_Init_ex" = xyes
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_HMAC_Init_ex" >&5
+printf "%s\n" "$ac_cv_lib_sunw_crypto_sunw_HMAC_Init_ex" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_HMAC_Init_ex" = xyes
 then :
 
       HAVECRYPTO="yes"
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lsunw_crypto $LIBS"
 else $as_nop
 
 
             { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking OpenSSL linking with -ldl" >&5
 printf %s "checking OpenSSL linking with -ldl... " >&6; }
-      LIBS="-lcrypto $CLEANLIBS -ldl"
+      LIBS="-lsunw_crypto $CLEANLIBS -ldl"
       cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -26674,7 +26674,7 @@
 int main (void)
 {
 
-        ERR_clear_error();
+        sunw_ERR_clear_error();
 
  ;
  return 0;
@@ -26694,7 +26694,7 @@
 
         { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking OpenSSL linking with -ldl and -lpthread" >&5
 printf %s "checking OpenSSL linking with -ldl and -lpthread... " >&6; }
-        LIBS="-lcrypto $CLEANLIBS -ldl -lpthread"
+        LIBS="-lsunw_crypto $CLEANLIBS -ldl -lpthread"
         cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -26704,7 +26704,7 @@
 int main (void)
 {
 
-          ERR_clear_error();
+          sunw_ERR_clear_error();
 
  ;
  return 0;
@@ -26741,14 +26741,14 @@
 
   if test X"$HAVECRYPTO" = X"yes"; then
 
-    { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for SSL_connect in -lssl" >&5
-printf %s "checking for SSL_connect in -lssl... " >&6; }
-if test ${ac_cv_lib_ssl_SSL_connect+y}
+    { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for sunw_SSL_connect in -lsunw_ssl" >&5
+printf %s "checking for sunw_SSL_connect in -lsunw_ssl... " >&6; }
+if test ${ac_cv_lib_sunw_ssl_sunw_SSL_connect+y}
 then :
   printf %s "(cached) " >&6
 else $as_nop
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lssl  $LIBS"
+LIBS="-lsunw_ssl  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -26756,10 +26756,10 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char SSL_connect ();
+char sunw_SSL_connect ();
 int main (void)
 {
-return SSL_connect ();
+return sunw_SSL_connect ();
  ;
  return 0;
 }
@@ -26766,38 +26766,38 @@
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
 then :
-  ac_cv_lib_ssl_SSL_connect=yes
+  ac_cv_lib_sunw_ssl_sunw_SSL_connect=yes
 else $as_nop
-  ac_cv_lib_ssl_SSL_connect=no
+  ac_cv_lib_sunw_ssl_sunw_SSL_connect=no
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.beam \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_ssl_SSL_connect" >&5
-printf "%s\n" "$ac_cv_lib_ssl_SSL_connect" >&6; }
-if test "x$ac_cv_lib_ssl_SSL_connect" = xyes
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_ssl_sunw_SSL_connect" >&5
+printf "%s\n" "$ac_cv_lib_sunw_ssl_sunw_SSL_connect" >&6; }
+if test "x$ac_cv_lib_sunw_ssl_sunw_SSL_connect" = xyes
 then :
   printf "%s\n" "#define HAVE_LIBSSL 1" >>confdefs.h
 
-  LIBS="-lssl $LIBS"
+  LIBS="-lsunw_ssl $LIBS"
 
 fi
 
 
-    if test "$ac_cv_lib_ssl_SSL_connect" != yes; then
+    if test "$ac_cv_lib_sunw_ssl_sunw_SSL_connect" != yes; then
             { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for ssl with RSAglue/rsaref libs in use" >&5
 printf %s "checking for ssl with RSAglue/rsaref libs in use... " >&6; };
       OLIBS=$LIBS
       LIBS="-lRSAglue -lrsaref $LIBS"
-      { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for SSL_connect in -lssl" >&5
-printf %s "checking for SSL_connect in -lssl... " >&6; }
-if test ${ac_cv_lib_ssl_SSL_connect+y}
+      { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for SSL_connect in -lsunw_ssl" >&5
+printf %s "checking for SSL_connect in -lsunw_ssl... " >&6; }
+if test ${ac_cv_lib_sunw_ssl_sunw_SSL_connect+y}
 then :
   printf %s "(cached) " >&6
 else $as_nop
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lssl  $LIBS"
+LIBS="-lsunw_ssl  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -26805,10 +26805,10 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char SSL_connect ();
+char sunw_SSL_connect ();
 int main (void)
 {
-return SSL_connect ();
+return sunw_SSL_connect ();
  ;
  return 0;
 }
@@ -26815,25 +26815,25 @@
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
 then :
-  ac_cv_lib_ssl_SSL_connect=yes
+  ac_cv_lib_sunw_ssl_sunw_SSL_connect=yes
 else $as_nop
-  ac_cv_lib_ssl_SSL_connect=no
+  ac_cv_lib_sunw_ssl_sunw_SSL_connect=no
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.beam \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_ssl_SSL_connect" >&5
-printf "%s\n" "$ac_cv_lib_ssl_SSL_connect" >&6; }
-if test "x$ac_cv_lib_ssl_SSL_connect" = xyes
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_ssl_sunw_SSL_connect" >&5
+printf "%s\n" "$ac_cv_lib_sunw_ssl_sunw_SSL_connect" >&6; }
+if test "x$ac_cv_lib_sunw_ssl_sunw_SSL_connect" = xyes
 then :
   printf "%s\n" "#define HAVE_LIBSSL 1" >>confdefs.h
 
-  LIBS="-lssl $LIBS"
+  LIBS="-lsunw_ssl $LIBS"
 
 fi
 
-      if test "$ac_cv_lib_ssl_SSL_connect" != yes; then
+      if test "$ac_cv_lib_sunw_ssl_sunw_SSL_connect" != yes; then
                 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: no" >&5
 printf "%s\n" "no" >&6; }
         LIBS=$OLIBS
