diff --git a/lib/http.js b/lib/http.js
index 3b1484b..0623668 100644
--- a/lib/http.js
+++ b/lib/http.js
@@ -344,9 +344,7 @@ IncomingMessage.prototype._read = function(n) {
   // We actually do almost nothing here, because the parserOnBody
   // function fills up our internal buffer directly.  However, we
   // do need to unpause the underlying socket so that it flows.
-  if (!this.socket.readable)
-    this.push(null);
-  else
+  if (this.socket.readable)
     readStart(this.socket);
 };
