diff -ru a/src/node_constants.cc b/src/node_constants.cc
--- a/src/node_constants.cc	2014-02-18 18:34:29.000000000 -0500
+++ b/src/node_constants.cc	2021-10-22 10:20:28.000000000 -0400
@@ -33,7 +33,7 @@
 #include <sys/stat.h>
 
 #if HAVE_OPENSSL
-# include <openssl/ssl.h>
+# include <openssl1x/ssl.h>
 #endif
 
 namespace node {
diff -ru a/src/node_crypto.h b/src/node_crypto.h
--- a/src/node_crypto.h	2014-02-18 18:34:29.000000000 -0500
+++ b/src/node_crypto.h	2021-10-22 10:20:17.000000000 -0400
@@ -27,15 +27,15 @@
 #include "node_object_wrap.h"
 #include "v8.h"
 
-#include <openssl/ssl.h>
-#include <openssl/err.h>
-#include <openssl/evp.h>
-#include <openssl/pem.h>
-#include <openssl/x509.h>
-#include <openssl/x509v3.h>
-#include <openssl/hmac.h>
-#include <openssl/rand.h>
-#include <openssl/pkcs12.h>
+#include <openssl1x/ssl.h>
+#include <openssl1x/err.h>
+#include <openssl1x/evp.h>
+#include <openssl1x/pem.h>
+#include <openssl1x/x509.h>
+#include <openssl1x/x509v3.h>
+#include <openssl1x/hmac.h>
+#include <openssl1x/rand.h>
+#include <openssl1x/pkcs12.h>
 
 #ifdef OPENSSL_NPN_NEGOTIATED
 #include "node_buffer.h"
