--- a/configure.orig	Wed Dec 18 23:49:45 2013
+++ b/configure	Mon Dec 23 21:03:16 2013
@@ -470,6 +470,7 @@
   target_arch = options.dest_cpu or host_arch
   o['variables']['host_arch'] = host_arch
   o['variables']['target_arch'] = target_arch
+  o['variables']['want_separate_host_toolset'] = '1'
 
   if target_arch == 'arm':
     configure_arm(o)
@@ -533,13 +534,13 @@
 
   # assume shared_zlib if one of these is set?
   if options.shared_zlib_libpath:
-    o['libraries'] += ['-L%s' % options.shared_zlib_libpath]
+    o['ldflags_target'] += ['-L%s' % options.shared_zlib_libpath]
   if options.shared_zlib_libname:
     o['libraries'] += ['-l%s' % options.shared_zlib_libname]
   elif options.shared_zlib:
     o['libraries'] += ['-lz']
   if options.shared_zlib_includes:
-    o['include_dirs'] += [options.shared_zlib_includes]
+    o['include_dirs_target'] += [options.shared_zlib_includes]
 
 
 def configure_http_parser(o):
@@ -590,13 +591,13 @@
 
   # assume shared_v8 if one of these is set?
   if options.shared_v8_libpath:
-    o['libraries'] += ['-L%s' % options.shared_v8_libpath]
+    o['ldflags_target'] += ['-L%s' % options.shared_v8_libpath]
   if options.shared_v8_libname:
     o['libraries'] += ['-l%s' % options.shared_v8_libname]
   elif options.shared_v8:
     o['libraries'] += ['-lv8']
   if options.shared_v8_includes:
-    o['include_dirs'] += [options.shared_v8_includes]
+    o['include_dirs_target'] += [options.shared_v8_includes]
 
 
 def configure_openssl(o):
@@ -613,7 +614,7 @@
     (libs, cflags) = pkg_config('openssl') or ('-lssl -lcrypto', '')
 
     if options.shared_openssl_libpath:
-      o['libraries'] += ['-L%s' % options.shared_openssl_libpath]
+      o['ldflags_target'] += ['-L%s' % options.shared_openssl_libpath]
 
     if options.shared_openssl_libname:
       libnames = options.shared_openssl_libname.split(',')
@@ -622,7 +623,7 @@
       o['libraries'] += libs.split()
 
     if options.shared_openssl_includes:
-      o['include_dirs'] += [options.shared_openssl_includes]
+      o['include_dirs_target'] += [options.shared_openssl_includes]
     else:
       o['cflags'] += cflags.split()
 
@@ -650,10 +651,11 @@
 
 output = {
   'variables': { 'python': sys.executable },
-  'include_dirs': [],
+  'include_dirs_target': [],
   'libraries': [],
   'defines': [],
   'cflags': [],
+  'ldflags_target': []
 }
 
 configure_node(output)
