diff -ru a/configure.sh b/configure.sh
--- a/configure.sh	Mon Dec 21 15:20:53 2015
+++ b/configure.sh	Thu Nov 21 16:13:44 2024
@@ -4479,6 +4479,7 @@
 
 #define _FILE_OFFSET_BITS 64
 #include <stdio.h>
+#include <stdlib.h>
 #include <fcntl.h>
 #include <sys/types.h>
 #include <sys/wait.h>
@@ -5685,6 +5686,7 @@
 #elif defined MAJOR_IN_SYSMACROS
 #include <sys/sysmacros.h>
 #endif
+#include <stdlib.h>
 
 int main(void)
 {
@@ -7903,7 +7905,7 @@
 #endif
 #include <stdlib.h>
 #include <errno.h>
-    main() {
+int    main() {
 	char const *dangling_symlink = "conftest.dangle";
 	unlink(dangling_symlink);
 	if (symlink("conftest.no-such", dangling_symlink) < 0) abort();
@@ -7947,7 +7949,7 @@
 #include <stdlib.h>
 #include <errno.h>
 #define FILENAME "conftest.dangle"
-    main() {
+int    main() {
 	unlink(FILENAME);
 	if (symlink("conftest.no-such", FILENAME) < 0) abort();
 	unlink(FILENAME "2");
@@ -7991,7 +7993,7 @@
 #include <stdlib.h>
 #include <errno.h>
 #define FILENAME "conftest.fifi"
-    main() {
+int    main() {
 	unlink(FILENAME);
 	if (mkfifo(FILENAME, 0777) < 0) abort();
 	unlink(FILENAME "2");
@@ -8031,8 +8033,9 @@
 
 #include <sys/types.h>
 #include <sys/socket.h>
+#include <stdlib.h>
 
-main() {
+int main() {
        int fd[2];
        exit((socketpair(AF_UNIX, SOCK_STREAM, 0, fd) != -1) ? 0 : 1);
 }
@@ -8264,7 +8267,8 @@
 /* end confdefs.h.  */
 #include <sys/types.h>
 #include <dirent.h>
-main() { struct dirent *di; DIR *d = opendir("."); di = readdir(d);
+#include <stdlib.h>
+int main() { struct dirent *di; DIR *d = opendir("."); di = readdir(d);
 if (di && di->d_name[-2] == '.' && di->d_name[-1] == 0 &&
 di->d_name[0] == 0) exit(0); exit(1);}
 _ACEOF
@@ -8296,6 +8300,7 @@
 /* end confdefs.h.  */
 #include <sys/types.h>
 #include <utime.h>
+#include <stdlib.h>
 int
 main ()
 {
@@ -8329,6 +8334,7 @@
 /* end confdefs.h.  */
 #include <sys/time.h>
 #include <unistd.h>
+#include <stdlib.h>
 int
 main ()
 {
@@ -8337,6 +8343,7 @@
   return 0;
 }
 _ACEOF
+cp conftest.$ac_ext /tmp/rsync-conftest-tz.$ac_ext
 if ac_fn_c_try_compile "$LINENO"; then :
   rsync_cv_HAVE_GETTIMEOFDAY_TZ=yes
 else
@@ -8366,6 +8373,7 @@
 
 #include <sys/types.h>
 #include <stdarg.h>
+#include <stdlib.h>
 void foo(const char *format, ...) {
        va_list ap;
        int len;
@@ -8380,7 +8388,7 @@
 
        exit(0);
 }
-main() { foo("hello"); }
+int main() { foo("hello"); }
 
 _ACEOF
 if ac_fn_c_try_run "$LINENO"; then :
@@ -8417,7 +8425,7 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <unistd.h>
-main() {
+int main() {
   struct stat st;
   char tpl[20]="/tmp/test.XXXXXX";
   int fd = mkstemp(tpl);
@@ -8470,7 +8478,7 @@
 #include <stdio.h>
 #include <sys/stat.h>
 #include <errno.h>
-main() { int rc, ec; char *fn = "fifo-test";
+int main() { int rc, ec; char *fn = "fifo-test";
 unlink(fn); rc = mknod(fn,S_IFIFO,0600); ec = errno; unlink(fn);
 if (rc) {printf("(%d %d) ",rc,ec); return ec;}
 return 0;}
@@ -8508,7 +8516,7 @@
 #include <stdio.h>
 #include <sys/stat.h>
 #include <errno.h>
-main() { int rc, ec; char *fn = "sock-test";
+int main() { int rc, ec; char *fn = "sock-test";
 unlink(fn); rc = mknod(fn,S_IFSOCK,0600); ec = errno; unlink(fn);
 if (rc) {printf("(%d %d) ",rc,ec); return ec;}
 return 0;}
Only in b: configure.sh~
