--- libntp/dolfptoa.c.orig	2021-01-17 11:12:05.000000000 -0800
+++ libntp/dolfptoa.c	2021-01-21 08:34:43.000000000 -0800
@@ -150,7 +150,7 @@
 		M_NEG(fpi, fpf);
 	}
 
-	return dolfptoa(fpi, fpf, (isneg?'-':'+'), ndec, FALSE);
+	return dolfptoa(fpi, fpf, (isneg?'-':'\0'), ndec, FALSE);
 }
 
 
@@ -168,7 +168,7 @@
 		M_NEG(fpi, fpf);
 	}
 
-	return dolfptoa(fpi, fpf, (isneg?'-':'+'), ndec, TRUE);
+	return dolfptoa(fpi, fpf, (isneg?'-':'\0'), ndec, TRUE);
 }
 
 
