--- include/ntp_config.h_orig	2020-12-15 09:59:29.608754176 +0000
+++ include/ntp_config.h	2020-12-15 10:00:02.755717774 +0000
@@ -15,7 +15,7 @@
  */
 #ifndef CONFIG_FILE
 # ifndef SYS_WINNT
-#  define	CONFIG_FILE "/etc/ntp.conf"
+#  define	CONFIG_FILE "/etc/inet/ntp.conf"
 # else /* SYS_WINNT */
 #  define	CONFIG_FILE	"%windir%\\system32\\drivers\\etc\\ntp.conf"
 #  define	ALT_CONFIG_FILE "%windir%\\ntp.conf"
--- libntp/audio.c_orig	2020-12-15 10:01:01.291870014 +0000
+++ libntp/audio.c	2020-12-15 10:01:31.639741992 +0000
@@ -66,7 +66,7 @@
 static struct audio_device device; /* audio device ident */
 #endif /* HAVE_SYS_AUDIOIO_H */
 #ifdef PCM_STYLE_SOUND
-# define INIT_FILE "/etc/ntp.audio"
+# define INIT_FILE "/etc/inet/ntp.audio"
 
 static ioctl_arg_T agc		= SOUND_MIXER_WRITE_RECLEV; /* or IGAIN or LINE */
 static ioctl_arg_T audiomonitor	= SOUND_MIXER_WRITE_VOLUME; /* or OGAIN */
--- ntpdate/ntpdate.c_orig	2020-12-15 10:02:37.046065626 +0000
+++ ntpdate/ntpdate.c	2020-12-15 10:03:04.849615327 +0000
@@ -179,7 +179,7 @@
 
 #ifndef KEYFILE
 # ifndef SYS_WINNT
-#define KEYFILE 	"/etc/ntp.keys"
+#define KEYFILE 	"/etc/inet/ntp.keys"
 # else
 #define KEYFILE 	"%windir%\\ntp.keys"
 # endif /* SYS_WINNT */
--- ntpd/refclock_oncore.c_orig	2020-12-15 10:03:48.326006011 +0000
+++ ntpd/refclock_oncore.c	2020-12-15 10:04:26.727103584 +0000
@@ -1213,7 +1213,7 @@
 
 	FILE	*fd;
 	char	*cc, *ca, line[100], units[2], device[64];
-	const char	*dirs[] = { "/etc/ntp", "/etc", 0 };
+	const char	*dirs[] = { "/etc/inet", "/etc", 0 };
 	const char *cp, **cpp;
 	int	i, sign, lat_flg, long_flg, ht_flg, mode, mask;
 	double	f1, f2, f3;
