--- libntp/audio.c	Tue Jul  1 18:15:20 2014
+++ libntp/audio.c.orig	Tue Jul  1 18:15:37 2014
@@ -55,7 +55,7 @@
 #ifdef PCM_STYLE_SOUND
 # define INIT_FILE "/etc/inet/ntp.audio"
 int agc =	SOUND_MIXER_WRITE_RECLEV; /* or IGAIN or LINE */
-int monitor =	SOUND_MIXER_WRITE_VOLUME; /* or OGAIN */
+int audiomonitor =	SOUND_MIXER_WRITE_VOLUME; /* or OGAIN */
 int devmask = 0;
 int recmask = 0;
 char cf_c_dev[100], cf_i_dev[100], cf_agc[100], cf_monitor[100];
@@ -334,7 +334,7 @@
 		/* devmask */
 		i = mixer_name(cf_monitor, devmask);
 		if (i >= 0)
-			monitor = MIXER_WRITE(i);
+			audiomonitor = MIXER_WRITE(i);
 		else
 			printf("monitor %s not in devmask %#x\n",
 			       cf_monitor, devmask);
@@ -409,7 +409,7 @@
 # endif
 		l |= r << 8;
 		if (cf_monitor[0] != '\0')
-			rval = ioctl(ctl_fd, monitor, &l );
+			rval = ioctl(ctl_fd, audiomonitor, &l );
 		else 
 			rval = ioctl(ctl_fd, SOUND_MIXER_WRITE_VOLUME,
 				     &l);
