--- scripts/lib/NTP/Util.pm	Wed Jul  2 20:25:10 2014
+++ scripts/lib/NTP/Util.pm	Wed Jul  2 22:37:15 2014
@@ -9,17 +9,9 @@
 my $ntpq_path = 'ntpq';
 my $sntp_path = 'sntp';
 
-our $IP_AGNOSTIC;
-
 BEGIN {
     require Socket;
-    if ($Socket::VERSION >= 1.94) {
-        Socket->import(qw(getaddrinfo getnameinfo SOCK_RAW AF_INET));
-        $IP_AGNOSTIC = 1;
-    }
-    else {
-        Socket->import(qw(inet_aton SOCK_RAW AF_INET));
-    }
+    Socket->import(qw(inet_aton SOCK_RAW AF_INET));
 }
 
 my %obsolete_vars = (
@@ -69,19 +61,7 @@
 sub do_dns {
     my ($host) = @_;
 
-    if ($IP_AGNOSTIC) {
-        my ($err, $res);
-
-        ($err, $res) = getaddrinfo($host, '', {socktype => SOCK_RAW});
-        die "getaddrinfo failed: $err\n" if $err;
-
-        ($err, $res) = getnameinfo($res->{addr}, 0);
-        die "getnameinfo failed: $err\n" if $err;
-
-        return $res;
-    }
-    # Too old perl, do only ipv4
-    elsif ($host =~ /^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})$/) {
+    if ($host =~ /^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})$/) {
         return gethostbyaddr inet_aton($host), AF_INET;
     }
     else {
