--- ipmitool-1.8.18-32/configure	2016-10-08 08:11:34.000000000 +0000
+++ ipmitool-1.8.18.N/configure	2020-04-15 14:44:51.484661703 +0000
@@ -12573,11 +12573,9 @@
 	LD=gld
 	AR=gar
 	# openssl libs are in /usr/sfw/lib on solaris 10
-	LIBS="$LIBS -R/usr/sfw/lib"
 	# disable the linux-specific interfaces
-	xenable_intf_bmc=yes
+	xenable_intf_bmc=no
 	xenable_intf_imb=no
-	xenable_intf_open=no
 	xenable_intf_lipmi=no
 	xenable_ipmishell=no
 	xenable_all_options=no
@@ -12676,13 +12674,13 @@
 fi
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for EVP_aes_128_cbc in -lcrypto" >&5
-$as_echo_n "checking for EVP_aes_128_cbc in -lcrypto... " >&6; }
-if ${ac_cv_lib_crypto_EVP_aes_128_cbc+:} false; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for sunw_EVP_aes_128_cbc in -lsunw_crypto" >&5
+$as_echo_n "checking for sunw_EVP_aes_128_cbc in -lsunw_crypto... " >&6; }
+if ${ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto -lcrypto $LIBS"
+LIBS="-lsunw_crypto -lsunw_crypto $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -12692,42 +12690,42 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char EVP_aes_128_cbc ();
+char sunw_EVP_aes_128_cbc ();
 int
 main ()
 {
-return EVP_aes_128_cbc ();
+return sunw_EVP_aes_128_cbc ();
   ;
   return 0;
 }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_lib_crypto_EVP_aes_128_cbc=yes
+  ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc=yes
 else
-  ac_cv_lib_crypto_EVP_aes_128_cbc=no
+  ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_EVP_aes_128_cbc" >&5
-$as_echo "$ac_cv_lib_crypto_EVP_aes_128_cbc" >&6; }
-if test "x$ac_cv_lib_crypto_EVP_aes_128_cbc" = xyes; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc" >&5
+$as_echo "$ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc" = xyes; then :
   if test "x$xenable_internal_md5" != "xyes"; then
-	    have_crypto=yes; LIBS="$LIBS -lcrypto"
+	    have_crypto=yes; LIBS="$LIBS -lsunw_crypto"
 	fi
 else
   have_crypto=no
 fi
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for EVP_sha256 in -lcrypto" >&5
-$as_echo_n "checking for EVP_sha256 in -lcrypto... " >&6; }
-if ${ac_cv_lib_crypto_EVP_sha256+:} false; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for sunw_EVP_sha256 in -lsunw_crypto" >&5
+$as_echo_n "checking for sunw_EVP_sha256 in -lsunw_crypto... " >&6; }
+if ${ac_cv_lib_sunw_crypto_sunw_EVP_sha256+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto -lcrypto $LIBS"
+LIBS="-lsunw_crypto -lsunw_crypto $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -12737,30 +12735,30 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char EVP_sha256 ();
+char sunw_EVP_sha256 ();
 int
 main ()
 {
-return EVP_sha256 ();
+return sunw_EVP_sha256 ();
   ;
   return 0;
 }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_lib_crypto_EVP_sha256=yes
+  ac_cv_lib_sunw_crypto_sunw_EVP_sha256=yes
 else
-  ac_cv_lib_crypto_EVP_sha256=no
+  ac_cv_lib_sunw_crypto_sunw_EVP_sha256=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_EVP_sha256" >&5
-$as_echo "$ac_cv_lib_crypto_EVP_sha256" >&6; }
-if test "x$ac_cv_lib_crypto_EVP_sha256" = xyes; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_EVP_sha256" >&5
+$as_echo "$ac_cv_lib_sunw_crypto_sunw_EVP_sha256" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_EVP_sha256" = xyes; then :
   if test "x$xenable_internal_sha256" != "xyes"; then
 		if test "x$have_crypto" != "xyes"; then
-			LIBS="$LIBS -lcrypto"
+			LIBS="$LIBS -lsunw_crypto"
 		have_sha256=yes
 		fi
 
@@ -12770,13 +12768,13 @@
 fi
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for MD5_Init in -lcrypto" >&5
-$as_echo_n "checking for MD5_Init in -lcrypto... " >&6; }
-if ${ac_cv_lib_crypto_MD5_Init+:} false; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for sunw_MD5_Init in -lsunw_crypto" >&5
+$as_echo_n "checking for sunw_MD5_Init in -lsunw_crypto... " >&6; }
+if ${ac_cv_lib_sunw_crypto_sunw_MD5_Init+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto -lcrypto $LIBS"
+LIBS="-lsunw_crypto -lsunw_crypto $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -12786,30 +12784,30 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char MD5_Init ();
+char sunw_MD5_Init ();
 int
 main ()
 {
-return MD5_Init ();
+return sunw_MD5_Init ();
   ;
   return 0;
 }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_lib_crypto_MD5_Init=yes
+  ac_cv_lib_sunw_crypto_sunw_MD5_Init=yes
 else
-  ac_cv_lib_crypto_MD5_Init=no
+  ac_cv_lib_sunw_crypto_sunw_MD5_Init=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_MD5_Init" >&5
-$as_echo "$ac_cv_lib_crypto_MD5_Init" >&6; }
-if test "x$ac_cv_lib_crypto_MD5_Init" = xyes; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_MD5_Init" >&5
+$as_echo "$ac_cv_lib_sunw_crypto_sunw_MD5_Init" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_MD5_Init" = xyes; then :
   if test "x$xenable_internal_md5" != "xyes"; then
 	    if test "x$have_crypto" != "xyes"; then
-	        LIBS="$LIBS -lcrypto"
+	        LIBS="$LIBS -lsunw_crypto"
 	        have_md5=yes
 	    fi
 
@@ -12819,13 +12817,13 @@
 fi
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for MD2_Init in -lcrypto" >&5
-$as_echo_n "checking for MD2_Init in -lcrypto... " >&6; }
-if ${ac_cv_lib_crypto_MD2_Init+:} false; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for sunw_MD2_Init in -lsunw_crypto" >&5
+$as_echo_n "checking for sunw_MD2_Init in -lsunw_crypto... " >&6; }
+if ${ac_cv_lib_sunw_crypto_sunw_MD2_Init+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto -lcrypto $LIBS"
+LIBS="-lsunw_crypto -lsunw_crypto $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -12835,30 +12833,30 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char MD2_Init ();
+char sunw_MD2_Init ();
 int
 main ()
 {
-return MD2_Init ();
+return sunw_MD2_Init ();
   ;
   return 0;
 }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_lib_crypto_MD2_Init=yes
+  ac_cv_lib_sunw_crypto_sunw_MD2_Init=yes
 else
-  ac_cv_lib_crypto_MD2_Init=no
+  ac_cv_lib_sunw_crypto_sunw_MD2_Init=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_MD2_Init" >&5
-$as_echo "$ac_cv_lib_crypto_MD2_Init" >&6; }
-if test "x$ac_cv_lib_crypto_MD2_Init" = xyes; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_MD2_Init" >&5
+$as_echo "$ac_cv_lib_sunw_crypto_sunw_MD2_Init" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_MD2_Init" = xyes; then :
   if test "x$xenable_internal_md5" != "xyes"; then
 	    if test "x$have_crypto" != "xyes" && test "x$have_md5" != "xyes"; then
-	        LIBS="$LIBS -lcrypto"
+	        LIBS="$LIBS -lsunw_crypto"
 	        have_md2=yes
 	    fi
 
