--- ipmitool-1.8.18-32/configure	2016-10-08 08:11:34.000000000 +0000
+++ ipmitool-1.8.18.N/configure	2019-06-27 15:41:38.646191188 +0000
@@ -2412,7 +2412,7 @@
     NONENONEs,x,x, &&
   program_prefix=${target_alias}-
 
-am__api_version='1.14'
+am__api_version='1.15'
 
 # Find a good install program.  We prefer a C program (faster),
 # so one script is as good as another.  But avoid the broken or
@@ -2604,7 +2604,7 @@
 $as_echo "$as_me: WARNING: 'missing' script is too old or missing" >&2;}
 fi
 
-if test x"${install_sh}" != xset; then
+if test x"${install_sh+set}" != xset; then
   case $am_aux_dir in
   *\ * | *\	*)
     install_sh="\${SHELL} '$am_aux_dir/install-sh'" ;;
@@ -2932,8 +2932,8 @@
 # <http://lists.gnu.org/archive/html/automake/2012-07/msg00014.html>
 mkdir_p='$(MKDIR_P)'
 
-# We need awk for the "check" target.  The system "awk" is bad on
-# some platforms.
+# We need awk for the "check" target (and possibly the TAP driver).  The
+# system "awk" is bad on some platforms.
 # Always define AMTAR for backward compatibility.  Yes, it's still used
 # in the wild :-(  We should find a proper way to deprecate it ...
 AMTAR='$${TAR-tar}'
@@ -5839,8 +5839,7 @@
     ;;
   *)
     lt_cv_sys_max_cmd_len=`(getconf ARG_MAX) 2> /dev/null`
-    if test -n "$lt_cv_sys_max_cmd_len" && \
-	test undefined != "$lt_cv_sys_max_cmd_len"; then
+    if test -n "$lt_cv_sys_max_cmd_len"; then
       lt_cv_sys_max_cmd_len=`expr $lt_cv_sys_max_cmd_len \/ 4`
       lt_cv_sys_max_cmd_len=`expr $lt_cv_sys_max_cmd_len \* 3`
     else
@@ -6241,6 +6240,10 @@
   fi
   ;;
 
+gnu*)
+  lt_cv_deplibs_check_method=pass_all
+  ;;
+
 haiku*)
   lt_cv_deplibs_check_method=pass_all
   ;;
@@ -6279,11 +6282,11 @@
   ;;
 
 # This must be glibc/ELF.
-linux* | k*bsd*-gnu | kopensolaris*-gnu | gnu*)
+linux* | k*bsd*-gnu | kopensolaris*-gnu)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
-netbsd* | netbsdelf*-gnu)
+netbsd*)
   if echo __ELF__ | $CC -E - | $GREP __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/]+(\.so\.[0-9]+\.[0-9]+|_pic\.a)$'
   else
@@ -7360,7 +7363,7 @@
   rm -rf conftest*
   ;;
 
-x86_64-*kfreebsd*-gnu|x86_64-*linux*|powerpc*-*linux*| \
+x86_64-*kfreebsd*-gnu|x86_64-*linux*|ppc*-*linux*|powerpc*-*linux*| \
 s390*-*linux*|s390*-*tpf*|sparc*-*linux*)
   # Find out which ABI we are using.
   echo 'int i;' > conftest.$ac_ext
@@ -7376,19 +7379,9 @@
 	    LD="${LD-ld} -m elf_i386_fbsd"
 	    ;;
 	  x86_64-*linux*)
-	    case `/usr/bin/file conftest.o` in
-	      *x86-64*)
-		LD="${LD-ld} -m elf32_x86_64"
-		;;
-	      *)
-		LD="${LD-ld} -m elf_i386"
-		;;
-	    esac
-	    ;;
-	  powerpc64le-*)
-	    LD="${LD-ld} -m elf32lppclinux"
+	    LD="${LD-ld} -m elf_i386"
 	    ;;
-	  powerpc64-*)
+	  ppc64-*linux*|powerpc64-*linux*)
 	    LD="${LD-ld} -m elf32ppclinux"
 	    ;;
 	  s390x-*linux*)
@@ -7407,10 +7400,7 @@
 	  x86_64-*linux*)
 	    LD="${LD-ld} -m elf_x86_64"
 	    ;;
-	  powerpcle-*)
-	    LD="${LD-ld} -m elf64lppc"
-	    ;;
-	  powerpc-*)
+	  ppc*-*linux*|powerpc*-*linux*)
 	    LD="${LD-ld} -m elf64ppc"
 	    ;;
 	  s390*-*linux*|s390*-*tpf*)
@@ -8947,7 +8937,7 @@
       lt_prog_compiler_static='-non_shared'
       ;;
 
-    linux* | k*bsd*-gnu | kopensolaris*-gnu | gnu*)
+    linux* | k*bsd*-gnu | kopensolaris*-gnu)
       case $cc_basename in
       # old Intel for x86_64 which still supported -KPIC.
       ecc*)
@@ -9425,9 +9415,6 @@
   openbsd*)
     with_gnu_ld=no
     ;;
-  linux* | k*bsd*-gnu | gnu*)
-    link_all_deplibs=no
-    ;;
   esac
 
   ld_shlibs=yes
@@ -9649,7 +9636,7 @@
       fi
       ;;
 
-    netbsd* | netbsdelf*-gnu)
+    netbsd*)
       if echo __ELF__ | $CC -E - | $GREP __ELF__ >/dev/null; then
 	archive_cmds='$LD -Bshareable $libobjs $deplibs $linker_flags -o $lib'
 	wlarc=
@@ -9826,7 +9813,6 @@
 	if test "$aix_use_runtimelinking" = yes; then
 	  shared_flag="$shared_flag "'${wl}-G'
 	fi
-	link_all_deplibs=no
       else
 	# not using gcc
 	if test "$host_cpu" = ia64; then
@@ -10280,7 +10266,7 @@
       link_all_deplibs=yes
       ;;
 
-    netbsd* | netbsdelf*-gnu)
+    netbsd*)
       if echo __ELF__ | $CC -E - | $GREP __ELF__ >/dev/null; then
 	archive_cmds='$LD -Bshareable -o $lib $libobjs $deplibs $linker_flags'  # a.out
       else
@@ -11117,6 +11103,17 @@
   esac
   ;;
 
+gnu*)
+  version_type=linux # correct to gnu/linux during the next big refactor
+  need_lib_prefix=no
+  need_version=no
+  library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${release}${shared_ext}${major} ${libname}${shared_ext}'
+  soname_spec='${libname}${release}${shared_ext}$major'
+  shlibpath_var=LD_LIBRARY_PATH
+  shlibpath_overrides_runpath=no
+  hardcode_into_libs=yes
+  ;;
+
 haiku*)
   version_type=linux # correct to gnu/linux during the next big refactor
   need_lib_prefix=no
@@ -11233,7 +11230,7 @@
   ;;
 
 # This must be glibc/ELF.
-linux* | k*bsd*-gnu | kopensolaris*-gnu | gnu*)
+linux* | k*bsd*-gnu | kopensolaris*-gnu)
   version_type=linux # correct to gnu/linux during the next big refactor
   need_lib_prefix=no
   need_version=no
@@ -11297,18 +11294,6 @@
   dynamic_linker='GNU/Linux ld.so'
   ;;
 
-netbsdelf*-gnu)
-  version_type=linux
-  need_lib_prefix=no
-  need_version=no
-  library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${release}${shared_ext}$major ${libname}${shared_ext}'
-  soname_spec='${libname}${release}${shared_ext}$major'
-  shlibpath_var=LD_LIBRARY_PATH
-  shlibpath_overrides_runpath=no
-  hardcode_into_libs=yes
-  dynamic_linker='NetBSD ld.elf_so'
-  ;;
-
 netbsd*)
   version_type=sunos
   need_lib_prefix=no
@@ -12573,11 +12558,9 @@
 	LD=gld
 	AR=gar
 	# openssl libs are in /usr/sfw/lib on solaris 10
-	LIBS="$LIBS -R/usr/sfw/lib"
 	# disable the linux-specific interfaces
-	xenable_intf_bmc=yes
+	xenable_intf_bmc=no
 	xenable_intf_imb=no
-	xenable_intf_open=no
 	xenable_intf_lipmi=no
 	xenable_ipmishell=no
 	xenable_all_options=no
@@ -12676,13 +12659,13 @@
 fi
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for EVP_aes_128_cbc in -lcrypto" >&5
-$as_echo_n "checking for EVP_aes_128_cbc in -lcrypto... " >&6; }
-if ${ac_cv_lib_crypto_EVP_aes_128_cbc+:} false; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for sunw_EVP_aes_128_cbc in -lsunw_crypto" >&5
+$as_echo_n "checking for sunw_EVP_aes_128_cbc in -lsunw_crypto... " >&6; }
+if ${ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto -lcrypto $LIBS"
+LIBS="-lsunw_crypto -lsunw_crypto $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -12692,42 +12675,42 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char EVP_aes_128_cbc ();
+char sunw_EVP_aes_128_cbc ();
 int
 main ()
 {
-return EVP_aes_128_cbc ();
+return sunw_EVP_aes_128_cbc ();
   ;
   return 0;
 }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_lib_crypto_EVP_aes_128_cbc=yes
+  ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc=yes
 else
-  ac_cv_lib_crypto_EVP_aes_128_cbc=no
+  ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_EVP_aes_128_cbc" >&5
-$as_echo "$ac_cv_lib_crypto_EVP_aes_128_cbc" >&6; }
-if test "x$ac_cv_lib_crypto_EVP_aes_128_cbc" = xyes; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc" >&5
+$as_echo "$ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_EVP_aes_128_cbc" = xyes; then :
   if test "x$xenable_internal_md5" != "xyes"; then
-	    have_crypto=yes; LIBS="$LIBS -lcrypto"
+	    have_crypto=yes; LIBS="$LIBS -lsunw_crypto"
 	fi
 else
   have_crypto=no
 fi
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for EVP_sha256 in -lcrypto" >&5
-$as_echo_n "checking for EVP_sha256 in -lcrypto... " >&6; }
-if ${ac_cv_lib_crypto_EVP_sha256+:} false; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for sunw_EVP_sha256 in -lsunw_crypto" >&5
+$as_echo_n "checking for sunw_EVP_sha256 in -lsunw_crypto... " >&6; }
+if ${ac_cv_lib_sunw_crypto_sunw_EVP_sha256+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto -lcrypto $LIBS"
+LIBS="-lsunw_crypto -lsunw_crypto $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -12737,30 +12720,30 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char EVP_sha256 ();
+char sunw_EVP_sha256 ();
 int
 main ()
 {
-return EVP_sha256 ();
+return sunw_EVP_sha256 ();
   ;
   return 0;
 }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_lib_crypto_EVP_sha256=yes
+  ac_cv_lib_sunw_crypto_sunw_EVP_sha256=yes
 else
-  ac_cv_lib_crypto_EVP_sha256=no
+  ac_cv_lib_sunw_crypto_sunw_EVP_sha256=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_EVP_sha256" >&5
-$as_echo "$ac_cv_lib_crypto_EVP_sha256" >&6; }
-if test "x$ac_cv_lib_crypto_EVP_sha256" = xyes; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_EVP_sha256" >&5
+$as_echo "$ac_cv_lib_sunw_crypto_sunw_EVP_sha256" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_EVP_sha256" = xyes; then :
   if test "x$xenable_internal_sha256" != "xyes"; then
 		if test "x$have_crypto" != "xyes"; then
-			LIBS="$LIBS -lcrypto"
+			LIBS="$LIBS -lsunw_crypto"
 		have_sha256=yes
 		fi
 
@@ -12770,13 +12753,13 @@
 fi
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for MD5_Init in -lcrypto" >&5
-$as_echo_n "checking for MD5_Init in -lcrypto... " >&6; }
-if ${ac_cv_lib_crypto_MD5_Init+:} false; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for sunw_MD5_Init in -lsunw_crypto" >&5
+$as_echo_n "checking for sunw_MD5_Init in -lsunw_crypto... " >&6; }
+if ${ac_cv_lib_sunw_crypto_sunw_MD5_Init+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto -lcrypto $LIBS"
+LIBS="-lsunw_crypto -lsunw_crypto $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -12786,30 +12769,30 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char MD5_Init ();
+char sunw_MD5_Init ();
 int
 main ()
 {
-return MD5_Init ();
+return sunw_MD5_Init ();
   ;
   return 0;
 }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_lib_crypto_MD5_Init=yes
+  ac_cv_lib_sunw_crypto_sunw_MD5_Init=yes
 else
-  ac_cv_lib_crypto_MD5_Init=no
+  ac_cv_lib_sunw_crypto_sunw_MD5_Init=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_MD5_Init" >&5
-$as_echo "$ac_cv_lib_crypto_MD5_Init" >&6; }
-if test "x$ac_cv_lib_crypto_MD5_Init" = xyes; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_MD5_Init" >&5
+$as_echo "$ac_cv_lib_sunw_crypto_sunw_MD5_Init" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_MD5_Init" = xyes; then :
   if test "x$xenable_internal_md5" != "xyes"; then
 	    if test "x$have_crypto" != "xyes"; then
-	        LIBS="$LIBS -lcrypto"
+	        LIBS="$LIBS -lsunw_crypto"
 	        have_md5=yes
 	    fi
 
@@ -12819,13 +12802,13 @@
 fi
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for MD2_Init in -lcrypto" >&5
-$as_echo_n "checking for MD2_Init in -lcrypto... " >&6; }
-if ${ac_cv_lib_crypto_MD2_Init+:} false; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for sunw_MD2_Init in -lsunw_crypto" >&5
+$as_echo_n "checking for sunw_MD2_Init in -lsunw_crypto... " >&6; }
+if ${ac_cv_lib_sunw_crypto_sunw_MD2_Init+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto -lcrypto $LIBS"
+LIBS="-lsunw_crypto -lsunw_crypto $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -12835,30 +12818,30 @@
 #ifdef __cplusplus
 extern "C"
 #endif
-char MD2_Init ();
+char sunw_MD2_Init ();
 int
 main ()
 {
-return MD2_Init ();
+return sunw_MD2_Init ();
   ;
   return 0;
 }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_lib_crypto_MD2_Init=yes
+  ac_cv_lib_sunw_crypto_sunw_MD2_Init=yes
 else
-  ac_cv_lib_crypto_MD2_Init=no
+  ac_cv_lib_sunw_crypto_sunw_MD2_Init=no
 fi
 rm -f core conftest.err conftest.$ac_objext \
     conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_crypto_MD2_Init" >&5
-$as_echo "$ac_cv_lib_crypto_MD2_Init" >&6; }
-if test "x$ac_cv_lib_crypto_MD2_Init" = xyes; then :
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_sunw_crypto_sunw_MD2_Init" >&5
+$as_echo "$ac_cv_lib_sunw_crypto_sunw_MD2_Init" >&6; }
+if test "x$ac_cv_lib_sunw_crypto_sunw_MD2_Init" = xyes; then :
   if test "x$xenable_internal_md5" != "xyes"; then
 	    if test "x$have_crypto" != "xyes" && test "x$have_md5" != "xyes"; then
-	        LIBS="$LIBS -lcrypto"
+	        LIBS="$LIBS -lsunw_crypto"
 	        have_md2=yes
 	    fi
 
--- ipmitool-1.8.18-32/configure.ac	2016-10-08 08:08:59.000000000 +0000
+++ ipmitool-1.8.18.N/configure.ac	2019-06-26 18:52:59.466303383 +0000
@@ -74,11 +74,9 @@
 	LD=gld
 	AR=gar
 	# openssl libs are in /usr/sfw/lib on solaris 10
-	LIBS="$LIBS -R/usr/sfw/lib"
 	# disable the linux-specific interfaces
-	xenable_intf_bmc=yes
+	xenable_intf_bmc=no
 	xenable_intf_imb=no
-	xenable_intf_open=no
 	xenable_intf_lipmi=no
 	xenable_ipmishell=no
 	xenable_all_options=no
@@ -154,41 +152,41 @@
 	[xenable_internal_md5=$enableval],
 	[xenable_internal_md5=no])
 
-AC_CHECK_LIB([crypto], [EVP_aes_128_cbc],
+AC_CHECK_LIB([sunw_crypto], [sunw_EVP_aes_128_cbc],
 	[if test "x$xenable_internal_md5" != "xyes"; then
-	    have_crypto=yes; LIBS="$LIBS -lcrypto"
+	    have_crypto=yes; LIBS="$LIBS -lsunw_crypto"
 	fi],
-	[have_crypto=no], [-lcrypto])
+	[have_crypto=no], [-lsunw_crypto])
 
-AC_CHECK_LIB([crypto], [EVP_sha256],
+AC_CHECK_LIB([sunw_crypto], [sunw_EVP_sha256],
 	[if test "x$xenable_internal_sha256" != "xyes"; then
 		if test "x$have_crypto" != "xyes"; then
-			LIBS="$LIBS -lcrypto"
+			LIBS="$LIBS -lsunw_crypto"
 		have_sha256=yes
 		fi
 		AC_DEFINE(HAVE_CRYPTO_SHA256, [1], [Define to 1 if libcrypto supports SHA256.])
 	fi],
-	[], [-lcrypto])
+	[], [-lsunw_crypto])
 
-AC_CHECK_LIB([crypto], [MD5_Init],
+AC_CHECK_LIB([sunw_crypto], [sunw_MD5_Init],
 	[if test "x$xenable_internal_md5" != "xyes"; then
 	    if test "x$have_crypto" != "xyes"; then
-	        LIBS="$LIBS -lcrypto"
+	        LIBS="$LIBS -lsunw_crypto"
 	        have_md5=yes
 	    fi
 	    AC_DEFINE(HAVE_CRYPTO_MD5, [1], [Define to 1 if libcrypto supports MD5.])
 	fi],
-	[], [-lcrypto])
+	[], [-lsunw_crypto])
 
-AC_CHECK_LIB([crypto], [MD2_Init],
+AC_CHECK_LIB([sunw_crypto], [sunw_MD2_Init],
 	[if test "x$xenable_internal_md5" != "xyes"; then
 	    if test "x$have_crypto" != "xyes" && test "x$have_md5" != "xyes"; then
-	        LIBS="$LIBS -lcrypto"
+	        LIBS="$LIBS -lsunw_crypto"
 	        have_md2=yes
 	    fi
 	    AC_DEFINE(HAVE_CRYPTO_MD2, [1], [Define to 1 if libcrypto supports MD2.])
 	fi],
-	[], [-lcrypto])
+	[], [-lsunw_crypto])
 
 dnl enable IPMIv1.5 LAN interface
 AC_ARG_ENABLE([intf-lan],
