diff -ru ipmitool-1.8.18-32/lib/helper.c ipmitool-1.8.18.N/lib/helper.c
--- ipmitool-1.8.18-32/lib/helper.c	2016-07-31 06:52:40.000000000 +0000
+++ ipmitool-1.8.18.N/lib/helper.c	2019-06-26 19:32:09.785252820 +0000
@@ -30,6 +30,7 @@
  * EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
  */
 #define _POSIX_SOURCE
+#define __EXTENSIONS__
 #define /* glibc 2.19 and earlier */ _BSD_SOURCE || \
 	/* Since glibc 2.20 */_DEFAULT_SOURCE || \
 	_XOPEN_SOURCE >= 500 || \
diff -ru ipmitool-1.8.18-32/lib/ipmi_hpmfwupg.c ipmitool-1.8.18.N/lib/ipmi_hpmfwupg.c
--- ipmitool-1.8.18-32/lib/ipmi_hpmfwupg.c	2016-06-29 18:01:49.000000000 +0000
+++ ipmitool-1.8.18.N/lib/ipmi_hpmfwupg.c	2019-06-26 19:07:01.800434943 +0000
@@ -46,6 +46,7 @@
 #include <stdio.h>
 #include <time.h>
 #include <sys/param.h>
+#include <sys/sysmacros.h>
 #include <unistd.h>
 
 #if HAVE_CONFIG_H
diff -ru ipmitool-1.8.18-32/lib/ipmi_isol.c ipmitool-1.8.18.N/lib/ipmi_isol.c
--- ipmitool-1.8.18-32/lib/ipmi_isol.c	2016-03-13 10:14:08.000000000 +0000
+++ ipmitool-1.8.18.N/lib/ipmi_isol.c	2019-06-27 15:47:22.231230537 +0000
@@ -30,6 +30,7 @@
  * EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
  */
 #define _XOPEN_SOURCE
+#define __EXTENSIONS__
 
 #include <stdlib.h>
 #include <string.h>
diff -ru ipmitool-1.8.18-32/lib/ipmi_main.c ipmitool-1.8.18.N/lib/ipmi_main.c
--- ipmitool-1.8.18-32/lib/ipmi_main.c	2016-07-31 06:56:33.000000000 +0000
+++ ipmitool-1.8.18.N/lib/ipmi_main.c	2019-06-27 15:49:29.001572468 +0000
@@ -34,6 +34,7 @@
 	(_XOPEN_SOURCE >= 500 || \
 	_XOPEN_SOURCE && _XOPEN_SOURCE_EXTENDED) && \
 	!(_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600)
+#define __EXTENSIONS__
 
 #include <stdlib.h>
 #include <stdio.h>
diff -ru ipmitool-1.8.18-32/lib/ipmi_sol.c ipmitool-1.8.18.N/lib/ipmi_sol.c
--- ipmitool-1.8.18-32/lib/ipmi_sol.c	2016-06-29 18:06:29.000000000 +0000
+++ ipmitool-1.8.18.N/lib/ipmi_sol.c	2019-06-27 15:47:54.602184146 +0000
@@ -34,6 +34,7 @@
 	(_XOPEN_SOURCE >= 500 || \
                        _XOPEN_SOURCE && _XOPEN_SOURCE_EXTENDED) && \
 	!(_POSIX_C_SOURCE >= 200809L || _XOPEN_SOURCE >= 700)
+#define __EXTENSIONS__
 
 #include <stdlib.h>
 #include <string.h>
diff -ru ipmitool-1.8.18-32/lib/ipmi_sunoem.c ipmitool-1.8.18.N/lib/ipmi_sunoem.c
--- ipmitool-1.8.18-32/lib/ipmi_sunoem.c	2016-05-30 18:28:40.000000000 +0000
+++ ipmitool-1.8.18.N/lib/ipmi_sunoem.c	2019-06-27 15:49:54.034008862 +0000
@@ -30,6 +30,7 @@
  * EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
  */
 #define _XOPEN_SOURCE
+#define __EXTENSIONS__
 
 #include <stdlib.h>
 #include <stdio.h>
