diff --git a/lib/ipmi_hpmfwupg.c b/lib/ipmi_hpmfwupg.c
index d63d2c1..03a790f 100644
--- a/lib/ipmi_hpmfwupg.c
+++ b/lib/ipmi_hpmfwupg.c
@@ -1,6 +1,7 @@
 /*
  * Copyright (c) 2006 Kontron Canada, Inc.  All Rights Reserved.
  * Copyright (c) 2003 Sun Microsystems, Inc.  All Rights Reserved.
+ * Copyright 2020 Joyent, Inc.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
@@ -48,6 +49,10 @@
 # include <config.h>
 #endif
 
+#ifndef MAX
+#define MAX(a,b) ((a) > (b) ? (a) : (b))
+#endif
+
 extern int verbose;
 
 VERSIONINFO gVersionInfo[HPMFWUPG_COMPONENT_ID_MAX];
