diff -ru ipmitool-1.8.18-32/src/plugins/ipmi_intf.c ipmitool-1.8.18.N/src/plugins/ipmi_intf.c
--- ipmitool-1.8.18-32/src/plugins/ipmi_intf.c	2016-07-31 06:52:40.000000000 +0000
+++ ipmitool-1.8.18.N/src/plugins/ipmi_intf.c	2019-06-27 15:50:11.662546324 +0000
@@ -439,10 +440,12 @@
 							 * byte. See also this page:
 							 * http://www.freebsd.org/doc/en/books/developers-handbook/ipv6.html
 							 */
+#if 0
 							if (IN6_IS_ADDR_LINKLOCAL(&tmp6->sin6_addr)
 									&& (tmp6->sin6_addr.s6_addr[1] != 0)) {
 								addr6->sin6_scope_id = ntohs(tmp6->sin6_addr.s6_addr[1]);
 							}
+#endif
 						}
 
 						/* OK, now try to connect with the scope id from this interface address */
