diff --git a/src/plugins/ipmi_intf.c b/src/plugins/ipmi_intf.c
index a84237e..621a235 100644
--- a/src/plugins/ipmi_intf.c
+++ b/src/plugins/ipmi_intf.c
@@ -477,10 +477,12 @@ ipmi_intf_socket_connect(struct ipmi_intf * intf)
 							 * byte. See also this page:
 							 * http://www.freebsd.org/doc/en/books/developers-handbook/ipv6.html
 							 */
+#if 0
 							if (IN6_IS_ADDR_LINKLOCAL(&tmp6->sin6_addr)
 									&& (tmp6->sin6_addr.s6_addr16[1] != 0)) {
 								addr6->sin6_scope_id = ntohs(tmp6->sin6_addr.s6_addr16[1]);
 							}
+#endif
 						}
 
 						/* OK, now try to connect with the scope id from this interface address */
